<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Spectre - simple discovery workflow • Spectre</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Spectre - simple discovery workflow">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Spectre</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-getting-started" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Getting started</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-getting-started">
<li><a class="dropdown-item" href="../articles/install.html">Install Spectre</a></li>
    <li><a class="dropdown-item" href="../articles/basics_guide.html">Basics guide</a></li>
    <li><a class="dropdown-item" href="../articles/data_prep.html">Preparing your data</a></li>
  </ul>
</li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-workflows" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Workflows</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-workflows">
<li><h6 class="dropdown-header" data-toc-skip>Core workflows</h6></li>
    <li><a class="dropdown-item" href="../articles/simple_discovery.html">Simple Discovery analysis</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Batch alignment workflows</h6></li>
    <li><a class="external-link dropdown-item" href="https://immunedynamics.io/spectre/cytonorm/">Batch alignment and analysis with CytoNorm</a></li>
    <li><a class="external-link dropdown-item" href="https://github.com/ImmuneDynamics/Spectre/blob/master/workflows/Spectre%20rPCA/Spectre%20rPCA.R">Batch alignment and analysis with rPCA</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials">
<li><a class="dropdown-item" href="../articles/data_transformation.html">Data transformation</a></li>
  </ul>
</li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ImmuneDynamics/Spectre/issues">Report an issue</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ImmuneDynamics/Spectre/discussions">Discussion &amp; FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ImmuneDynamics/Spectre/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Spectre - simple discovery workflow</h1>
                        <h4 data-toc-skip class="author">Thomas
Ashhurst, Givanna Putri</h4>
            
            <h4 data-toc-skip class="date">2025-09-19</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ImmuneDynamics/Spectre/blob/master/vignettes/simple_discovery.Rmd" class="external-link"><code>vignettes/simple_discovery.Rmd</code></a></small>
      <div class="d-none name"><code>simple_discovery.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><strong>Overview</strong></p>
<p>Here we provide a worked example of a ‘simple’ discovery analysis
workflow, where the entire process (data prep, clustering,
dimensionality reduction, cluster annotation, plotting, summary data,
and statistical analysis) is contained within a single script. The
‘simple’ workflow is most suitable for fast analysis of small datasets.
For larger or more complex datasets, or datasets with multiple batches,
we recommend the general discovery workflow, where the data preparation,
batch alignment, clustering/dimensionality reduction, and quantitative
analysis are separated into separate scripts.</p>
<p>The demo dataset used for this worked example are cells extracted
from mock- or virally-infected mouse brains, measured by flow
cytometry.</p>
<p><strong>Strategy</strong></p>
<p>The ‘simple’ and ‘general’ discovery workflows are designed to
facilitate the analysis of large and complex cytometry datasets using
the Spectre R package. We’ve tested up to 30 million cells in a single
analysis session so far. The workflow is designed to get around the cell
number limitations of tSNE/UMAP. The analysis starts with clustering
with FlowSOM – which is fast and scales well to large datasets. The
clustered data is then downsampled, and dimensionality reduction is
performed with tSNE/UMAP. This allows for visualisation of the data, and
the clusters present in the dataset. Once the possible cell types in the
datasets have been explored, the clusters can be labelled with the
appropriate cellular identities. Finally, we can use the
clusters/populations to generate summary statistics (expression levels,
frequencies, total counts etc), which allows us to create graphs and
heatmaps, facilitating statistical analysis.</p>
<p><strong>Multiple samples</strong></p>
<p>To analyse multiple samples, all the files must be imported into the
one analysis session. This allows cells from each session to be
clustered and analysed together, and allows us to examine the
differential expression of markers, or the differences in cell
proportions, between experimental groups.</p>
<p><strong>Batch alignment</strong></p>
<p>The ‘simple’ discovery workflow does not include any batch alignment
steps. If batch correction needs to be applied, we recommend using the
general discovery workflow.</p>
<p><img src="images/simple_discovery/image2021-2-28_11-47-29.png" style="width:70.0%"></p>
<p><img src="images/simple_discovery/image2021-2-28_11-49-51.png" style="width:70.0%"></p>
</div>
<div class="section level2">
<h2 id="before-you-start">Before you start<a class="anchor" aria-label="anchor" href="#before-you-start"></a>
</h2>
<p>If you haven’t installed Spectre, please visit our <a href="https://immunedynamics.io/spectre/getting-started/" class="external-link">Spectre
installation</a> page. If you are new to R and Spectre, we recommend
trying out the R/RStudio and Spectre <a href="https://immunedynamics.io/spectre/getting-started/" class="external-link">tutorials</a>
available on our <a href="https://immunedynamics.io/spectre/getting-started/" class="external-link">getting
started</a> page, to familiarise yourself with R/RStudio first.</p>
</div>
<div class="section level2">
<h2 id="citation-and-methods">Citation and methods<a class="anchor" aria-label="anchor" href="#citation-and-methods"></a>
</h2>
<p><strong>Citation</strong></p>
<p>If you use Spectre in your work, please consider citing <a href="https://onlinelibrary.wiley.com/doi/full/10.1002/cyto.a.24350" class="external-link">Ashhurst
TM, Marsh-Wakefield F, Putri GH et al. (2022). Cytometry Part A 101 (3),
237-253</a>. To continue providing open-source tools such as Spectre, it
helps us if we can demonstrate that our efforts are contributing to
analysis efforts in the community. Please also consider citing the
authors of the individual packages or tools (e.g. CytoNorm, FlowSOM,
tSNE, UMAP, etc) that are critical elements of your analysis work. We
have provided some generic text that you can use for your methods
section with each protocol and on the ‘about’ page.</p>
<p><strong>Sample methods blurb</strong></p>
<p>Here is a sample methods blurb for this workflow. You may need to
adapt this text to reflect any changes made in your analysis.</p>
<p><em>Computational analysis of data was performed using the Spectre R
package (Ashhurst et al., 2022), with instructions and source code
provided at <a href="https://github.com/ImmuneDynamics/spectre" class="external-link uri">https://github.com/ImmuneDynamics/spectre</a>. Samples were
initially prepared in FlowJo, and the population of interest was
exported as raw value CSV files. Arcsinh transformation was performed on
the data in R using a co-factor of 15 to redistribute the data on a
linear scale and compress low end values near zero. The dataset was then
merged into a single data.table, with keywords denoting the sample,
group, and other factors added to each row (cell). The FlowSOM algorithm
(Van Gassen et al., 2015) was then run on the merged dataset to cluster
the data, where every cell is assigned to a specific cluster and
metacluster. Subsequently, the data was downsampled and analysed by the
dimensionality reduction algorithm Uniform Manifold Approximation and
Projection (UMAP) (McInnes, Healy, Melville, 2018) for cellular
visualisation.</em></p>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<div class="section level4">
<h4 id="directories">Directories<a class="anchor" aria-label="anchor" href="#directories"></a>
</h4>
<p>Create a master folder with a meaningful name. Then inside that
folder, insert the following:</p>
<ul>
<li>One folder called ‘data’ – this will contain your data CSV or FCS
files</li>
<li>One folder called ‘metadata’ – this will contain a CSV containg your
sample metadata</li>
<li>One folder called ‘Spectre simple discovery’ or similar – place this
analysis script there</li>
</ul>
<p>Example:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="sc">/</span>spectre_simple_discovery  <span class="co"># Master folder for this simple discovery)}</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>  <span class="sc">/</span>data</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>    <span class="sc">--</span> Contains data files, one CSV or FCS per sample</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>  <span class="sc">/</span>metadata</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>    <span class="sc">--</span> Contains a CSV containing sample <span class="fu">metadata</span> (group, batch, etc)</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>  <span class="sc">/</span>scripts</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>    <span class="sc">--</span> Spectre simple.discovery.R</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="analysis-script">Analysis script<a class="anchor" aria-label="anchor" href="#analysis-script"></a>
</h4>
<p>You can download the <strong>simple discovery</strong> script from <a href="https://github.com/ImmuneDynamics/Spectre/tree/master/workflows" class="external-link">this
link</a>. Place this inside the <code>scripts</code> folder.</p>
</div>
<div class="section level4">
<h4 id="data-files">Data files<a class="anchor" aria-label="anchor" href="#data-files"></a>
</h4>
<p>If you would to use the <strong>demo data</strong> as a test run for
the simple discovery workflow, nothing to do at this step. Simply follow
the relevant instructions further down this page to download the demo
data.</p>
<p>If you would like to use your <strong>own data</strong>, add your
data and metadata files:</p>
<ul>
<li>Place the sample CSV or FCS files in the <code>data</code> folder
you created above.</li>
<li>Place the metadata CSV files in the <code>metadata</code> folder you
created above.</li>
</ul>
<!-- Please see [this page](http://immunedynamics.io/Spectre/) for detailed instructions on exporting data for Spectre and setting up a metadata file.  -->
</div>
</div>
<div class="section level2">
<h2 id="load-packages-and-set-directories">1. Load packages and set directories<a class="anchor" aria-label="anchor" href="#load-packages-and-set-directories"></a>
</h2>
<p>Running <code><a href="https://github.com/ImmuneDynamics/Spectre" class="external-link">library(Spectre)</a></code> will load the Spectre package
(also known as a ‘library’). We can then use
<code><a href="../reference/package.check.html">package.check()</a></code> to see if the standard dependency packages
are installed, and <code><a href="../reference/package.load.html">package.load()</a></code> to load those
packages.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Load libraries</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ImmuneDynamics/Spectre" class="external-link">Spectre</a></span><span class="op">)</span></span>
<span><span class="co"># Check that all required packages are installed</span></span>
<span><span class="fu">Spectre</span><span class="fu">::</span><span class="fu"><a href="../reference/package.check.html">package.check</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># Load required packages</span></span>
<span><span class="fu">Spectre</span><span class="fu">::</span><span class="fu"><a href="../reference/package.load.html">package.load</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Define a few directories to store data and results.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PrimaryDirectory</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"spectre_simple_discovery"</span><span class="op">)</span> <span class="co"># For demo purposes, we use a temporary directory</span></span>
<span><span class="co"># PrimaryDirectory &lt;- "C:/Path/To/Your/Directory/spectre_simple_discovery" # Uncomment and edit this line to set your own directory</span></span>
<span><span class="va">InputDirectory</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PrimaryDirectory</span>, <span class="st">"data"</span><span class="op">)</span> <span class="co"># Directory containing data files</span></span>
<span><span class="va">MetaDirectory</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PrimaryDirectory</span>, <span class="st">"metadata"</span><span class="op">)</span> <span class="co"># Directory containing metadata files</span></span>
<span><span class="va">OutputDirectory</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PrimaryDirectory</span>, <span class="st">"output"</span><span class="op">)</span> <span class="co"># Directory to save results</span></span></code></pre></div>
<p>Create them.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">PrimaryDirectory</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">InputDirectory</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">MetaDirectory</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">OutputDirectory</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="import-and-prep-data">2. Import and prep data<a class="anchor" aria-label="anchor" href="#import-and-prep-data"></a>
</h2>
<div class="section level3">
<h3 id="demo-data">Demo data<a class="anchor" aria-label="anchor" href="#demo-data"></a>
</h3>
<p>If you need the demo dataset, run the following code in the analysis
script (select all, CMD+SHIFT+C) and run to download. <strong>If you are
using your own datasets, then skip this step.</strong> Alternative,
download the files from <a href="https://github.com/ImmuneDynamics/data/blob/main/msCNS.zip?raw=TRUE" class="external-link">https://github.com/ImmuneDynamics/data/blob/main/msCNS.zip?raw=TRUE</a>.</p>
<p>This code will download the demo dataset files and metadata file, and
place them in the <code>data</code> and <code>metadata</code> folders
respectively.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://github.com/ImmuneDynamics/data/blob/main/msCNS.zip?raw=TRUE"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PrimaryDirectory</span>, <span class="st">'msCNS.zip'</span><span class="op">)</span>, </span>
<span>    mode <span class="op">=</span> <span class="st">'wb'</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span>zipfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PrimaryDirectory</span>, <span class="st">'msCNS.zip'</span><span class="op">)</span>, exdir <span class="op">=</span> <span class="va">PrimaryDirectory</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PrimaryDirectory</span>, <span class="st">'msCNS'</span>, <span class="st">'data'</span><span class="op">)</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.rename</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">i</span>,  to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">InputDirectory</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PrimaryDirectory</span>, <span class="st">'msCNS'</span>, <span class="st">'metadata'</span><span class="op">)</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.rename</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">i</span>,  to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">MetaDirectory</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlink.html" class="external-link">unlink</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PrimaryDirectory</span>, <span class="st">'msCNS'</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">PrimaryDirectory</span>, <span class="st">'msCNS.zip'</span><span class="op">)</span><span class="op">)</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="import-data">Import data<a class="anchor" aria-label="anchor" href="#import-data"></a>
</h3>
<p>To begin, we will change our working directory to ‘InputDirectory’
and list all the CSV files in that directory – these should be the
sample CSV files. We can then read in all of our samples (in this
example, one CSV file per sample) into a list called ‘data.list’.
Spectre uses the data.table framework to store data, which reads,
writes, and performs operations on data very quickly.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Import data</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">InputDirectory</span>, <span class="st">".csv"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "CNS_Mock_01.csv"   "CNS_Mock_02.csv"   "CNS_Mock_03.csv"  </span></span>
<span><span class="co">#&gt;  [4] "CNS_Mock_04.csv"   "CNS_Mock_05.csv"   "CNS_Mock_06.csv"  </span></span>
<span><span class="co">#&gt;  [7] "CNS_WNV_D7_01.csv" "CNS_WNV_D7_02.csv" "CNS_WNV_D7_03.csv"</span></span>
<span><span class="co">#&gt; [10] "CNS_WNV_D7_04.csv" "CNS_WNV_D7_05.csv" "CNS_WNV_D7_06.csv"</span></span>
<span></span>
<span><span class="va">data.list</span> <span class="op">&lt;-</span> <span class="fu">Spectre</span><span class="fu">::</span><span class="fu"><a href="../reference/read.files.html">read.files</a></span><span class="op">(</span></span>
<span>    file.loc <span class="op">=</span> <span class="va">InputDirectory</span>,</span>
<span>    file.type <span class="op">=</span> <span class="st">".csv"</span>,</span>
<span>    do.embed.file.names <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Reading CSV files...</span></span>
<span><span class="co">#&gt; Reading CNS_Mock_01.csv</span></span>
<span><span class="co">#&gt; Reading CNS_Mock_02.csv</span></span>
<span><span class="co">#&gt; Reading CNS_Mock_03.csv</span></span>
<span><span class="co">#&gt; Reading CNS_Mock_04.csv</span></span>
<span><span class="co">#&gt; Reading CNS_Mock_05.csv</span></span>
<span><span class="co">#&gt; Reading CNS_Mock_06.csv</span></span>
<span><span class="co">#&gt; Reading CNS_WNV_D7_01.csv</span></span>
<span><span class="co">#&gt; Reading CNS_WNV_D7_02.csv</span></span>
<span><span class="co">#&gt; Reading CNS_WNV_D7_03.csv</span></span>
<span><span class="co">#&gt; Reading CNS_WNV_D7_04.csv</span></span>
<span><span class="co">#&gt; Reading CNS_WNV_D7_05.csv</span></span>
<span><span class="co">#&gt; Reading CNS_WNV_D7_06.csv</span></span></code></pre></div>
<p>By default, the <code><a href="../reference/read.files.html">read.files()</a></code> function will generate some
other variables, which you can review, by running the
<code><a href="../reference/do.list.summary.html">do.list.summary()</a></code> function.</p>
<p>The <code>name.table</code> variable is a table of all the column
names for all of your samples (one row per sample, one column per column
name). If all of the column names are matching, then this table should
be a repeating pattern. If it has been jumbled, then some of your
samples have columns that don’t appear in other samples. The
<code>ncol.check</code> and <code>nrow.check</code> are simple tables
indicating the number or columns and rows in each sample.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Check the data</span></span>
<span></span>
<span><span class="va">check</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do.list.summary.html">do.list.summary</a></span><span class="op">(</span><span class="va">data.list</span><span class="op">)</span></span>
<span></span>
<span><span class="va">check</span><span class="op">$</span><span class="va">name.table</span> <span class="co"># Review column names and their subsequent values</span></span>
<span><span class="co">#&gt;      X1  X2   X3   X4    X5   X6   X7   X8  X9      X10    X11</span></span>
<span><span class="co">#&gt; 1  NK11 CD3 CD45 Ly6G CD11b B220 CD8a Ly6C CD4 FileName FileNo</span></span>
<span><span class="co">#&gt; 2  NK11 CD3 CD45 Ly6G CD11b B220 CD8a Ly6C CD4 FileName FileNo</span></span>
<span><span class="co">#&gt; 3  NK11 CD3 CD45 Ly6G CD11b B220 CD8a Ly6C CD4 FileName FileNo</span></span>
<span><span class="co">#&gt; 4  NK11 CD3 CD45 Ly6G CD11b B220 CD8a Ly6C CD4 FileName FileNo</span></span>
<span><span class="co">#&gt; 5  NK11 CD3 CD45 Ly6G CD11b B220 CD8a Ly6C CD4 FileName FileNo</span></span>
<span><span class="co">#&gt; 6  NK11 CD3 CD45 Ly6G CD11b B220 CD8a Ly6C CD4 FileName FileNo</span></span>
<span><span class="co">#&gt; 7  NK11 CD3 CD45 Ly6G CD11b B220 CD8a Ly6C CD4 FileName FileNo</span></span>
<span><span class="co">#&gt; 8  NK11 CD3 CD45 Ly6G CD11b B220 CD8a Ly6C CD4 FileName FileNo</span></span>
<span><span class="co">#&gt; 9  NK11 CD3 CD45 Ly6G CD11b B220 CD8a Ly6C CD4 FileName FileNo</span></span>
<span><span class="co">#&gt; 10 NK11 CD3 CD45 Ly6G CD11b B220 CD8a Ly6C CD4 FileName FileNo</span></span>
<span><span class="co">#&gt; 11 NK11 CD3 CD45 Ly6G CD11b B220 CD8a Ly6C CD4 FileName FileNo</span></span>
<span><span class="co">#&gt; 12 NK11 CD3 CD45 Ly6G CD11b B220 CD8a Ly6C CD4 FileName FileNo</span></span>
<span><span class="va">check</span><span class="op">$</span><span class="va">ncol.check</span> <span class="co"># Review number of columns (features, markers) in each sample</span></span>
<span><span class="co">#&gt;       [,1]</span></span>
<span><span class="co">#&gt;  [1,] 11  </span></span>
<span><span class="co">#&gt;  [2,] 11  </span></span>
<span><span class="co">#&gt;  [3,] 11  </span></span>
<span><span class="co">#&gt;  [4,] 11  </span></span>
<span><span class="co">#&gt;  [5,] 11  </span></span>
<span><span class="co">#&gt;  [6,] 11  </span></span>
<span><span class="co">#&gt;  [7,] 11  </span></span>
<span><span class="co">#&gt;  [8,] 11  </span></span>
<span><span class="co">#&gt;  [9,] 11  </span></span>
<span><span class="co">#&gt; [10,] 11  </span></span>
<span><span class="co">#&gt; [11,] 11  </span></span>
<span><span class="co">#&gt; [12,] 11</span></span>
<span><span class="va">check</span><span class="op">$</span><span class="va">nrow.check</span> <span class="co"># Review number of rows (cells) in each sample</span></span>
<span><span class="co">#&gt;       [,1] </span></span>
<span><span class="co">#&gt;  [1,] 9937 </span></span>
<span><span class="co">#&gt;  [2,] 15415</span></span>
<span><span class="co">#&gt;  [3,] 14246</span></span>
<span><span class="co">#&gt;  [4,] 17044</span></span>
<span><span class="co">#&gt;  [5,] 5459 </span></span>
<span><span class="co">#&gt;  [6,] 4891 </span></span>
<span><span class="co">#&gt;  [7,] 17950</span></span>
<span><span class="co">#&gt;  [8,] 16233</span></span>
<span><span class="co">#&gt;  [9,] 15999</span></span>
<span><span class="co">#&gt; [10,] 17131</span></span>
<span><span class="co">#&gt; [11,] 15926</span></span>
<span><span class="co">#&gt; [12,] 18773</span></span></code></pre></div>
<p>You can review the first 6 rows of the first sample in your data
using the following:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data.list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt;            NK11        CD3     CD45       Ly6G     CD11b      B220      CD8a</span></span>
<span><span class="co">#&gt;           &lt;num&gt;      &lt;num&gt;    &lt;num&gt;      &lt;num&gt;     &lt;num&gt;     &lt;num&gt;     &lt;num&gt;</span></span>
<span><span class="co">#&gt;    1:   42.3719  40.098700  6885.08  -344.7830 14787.300  -40.2399  83.71750</span></span>
<span><span class="co">#&gt;    2:   42.9586 119.014000  1780.29  -429.6650  5665.730   86.6673  34.72190</span></span>
<span><span class="co">#&gt;    3:   59.2366 206.238000 10248.30 -1603.8400 19894.300  427.8310 285.88000</span></span>
<span><span class="co">#&gt;    4:  364.9480  -0.233878  3740.04  -815.9800  9509.430  182.4200 333.60500</span></span>
<span><span class="co">#&gt;    5:  440.2470  40.035200  9191.38    40.5055  5745.820 -211.6940 149.22000</span></span>
<span><span class="co">#&gt;   ---                                                                       </span></span>
<span><span class="co">#&gt; 9933:   11.2126  36.951600  2515.82  -647.4930  6172.070  221.9380 266.90000</span></span>
<span><span class="co">#&gt; 9934:  239.9700 440.217000  7247.28 -1449.7200 15355.400  809.3040 456.59900</span></span>
<span><span class="co">#&gt; 9935: -134.9650 111.350000  2472.85    81.5975  9657.160 -113.1320   3.79607</span></span>
<span><span class="co">#&gt; 9936:   86.3333  28.286900  5745.27 -1284.0800 18303.100  353.5290 262.96300</span></span>
<span><span class="co">#&gt; 9937:   10.1467 122.255000  1971.69  -215.7660   727.708  506.8580 113.14400</span></span>
<span><span class="co">#&gt;            Ly6C      CD4    FileName FileNo</span></span>
<span><span class="co">#&gt;           &lt;num&gt;    &lt;num&gt;      &lt;char&gt;  &lt;int&gt;</span></span>
<span><span class="co">#&gt;    1:  958.7000  711.072 CNS_Mock_01      1</span></span>
<span><span class="co">#&gt;    2:  448.2590  307.272 CNS_Mock_01      1</span></span>
<span><span class="co">#&gt;    3: 1008.8300  707.094 CNS_Mock_01      1</span></span>
<span><span class="co">#&gt;    4:  440.0710  249.784 CNS_Mock_01      1</span></span>
<span><span class="co">#&gt;    5:   87.4815  867.570 CNS_Mock_01      1</span></span>
<span><span class="co">#&gt;   ---                                      </span></span>
<span><span class="co">#&gt; 9933:  141.4200  708.348 CNS_Mock_01      1</span></span>
<span><span class="co">#&gt; 9934: 2093.6900 2119.270 CNS_Mock_01      1</span></span>
<span><span class="co">#&gt; 9935: -114.1510  110.743 CNS_Mock_01      1</span></span>
<span><span class="co">#&gt; 9936:  745.8080  537.750 CNS_Mock_01      1</span></span>
<span><span class="co">#&gt; 9937:  244.2210 2334.800 CNS_Mock_01      1</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="merge-data-tables">Merge data.tables<a class="anchor" aria-label="anchor" href="#merge-data-tables"></a>
</h3>
<p>Once the metadata has been added, we can then merge the data into a
single data.table using do.merge.files(). By default, columns with
matching names will be aligned in the new table, and any columns that
are present in some samples, but not others, will be added and filled
with ‘NA’ for any samples that didn’t have that column initially. Once
the data has been merged, we can review the data:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Merge data</span></span>
<span></span>
<span><span class="va">cell.dat</span> <span class="op">&lt;-</span> <span class="fu">Spectre</span><span class="fu">::</span><span class="fu"><a href="../reference/do.merge.files.html">do.merge.files</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">data.list</span><span class="op">)</span></span>
<span><span class="va">cell.dat</span></span>
<span><span class="co">#&gt;              NK11        CD3     CD45       Ly6G    CD11b      B220      CD8a</span></span>
<span><span class="co">#&gt;             &lt;num&gt;      &lt;num&gt;    &lt;num&gt;      &lt;num&gt;    &lt;num&gt;     &lt;num&gt;     &lt;num&gt;</span></span>
<span><span class="co">#&gt;      1:   42.3719  40.098700  6885.08  -344.7830 14787.30  -40.2399   83.7175</span></span>
<span><span class="co">#&gt;      2:   42.9586 119.014000  1780.29  -429.6650  5665.73   86.6673   34.7219</span></span>
<span><span class="co">#&gt;      3:   59.2366 206.238000 10248.30 -1603.8400 19894.30  427.8310  285.8800</span></span>
<span><span class="co">#&gt;      4:  364.9480  -0.233878  3740.04  -815.9800  9509.43  182.4200  333.6050</span></span>
<span><span class="co">#&gt;      5:  440.2470  40.035200  9191.38    40.5055  5745.82 -211.6940  149.2200</span></span>
<span><span class="co">#&gt;     ---                                                                      </span></span>
<span><span class="co">#&gt; 169000:  910.8890  72.856100 31466.20  -316.5570 28467.80   -7.7972 -271.8040</span></span>
<span><span class="co">#&gt; 169001:  -10.2642  64.188700 45188.00  -540.5140 22734.00  202.4110 -936.4920</span></span>
<span><span class="co">#&gt; 169002: -184.2910  -9.445650 11842.60   -97.9383 17237.00  123.4760 -219.9320</span></span>
<span><span class="co">#&gt; 169003:  248.3860 229.986000 32288.20  -681.1630 19255.80 -656.0540 -201.5880</span></span>
<span><span class="co">#&gt; 169004:  738.9810  95.470300 46185.10 -1004.6000 22957.80 -661.6280   72.3356</span></span>
<span><span class="co">#&gt;               Ly6C       CD4      FileName FileNo</span></span>
<span><span class="co">#&gt;              &lt;num&gt;     &lt;num&gt;        &lt;char&gt;  &lt;int&gt;</span></span>
<span><span class="co">#&gt;      1:   958.7000  711.0720   CNS_Mock_01      1</span></span>
<span><span class="co">#&gt;      2:   448.2590  307.2720   CNS_Mock_01      1</span></span>
<span><span class="co">#&gt;      3:  1008.8300  707.0940   CNS_Mock_01      1</span></span>
<span><span class="co">#&gt;      4:   440.0710  249.7840   CNS_Mock_01      1</span></span>
<span><span class="co">#&gt;      5:    87.4815  867.5700   CNS_Mock_01      1</span></span>
<span><span class="co">#&gt;     ---                                          </span></span>
<span><span class="co">#&gt; 169000: 12023.7000 1103.0500 CNS_WNV_D7_06     12</span></span>
<span><span class="co">#&gt; 169001:  4188.3300  315.9400 CNS_WNV_D7_06     12</span></span>
<span><span class="co">#&gt; 169002:  8923.4000 -453.4640 CNS_WNV_D7_06     12</span></span>
<span><span class="co">#&gt; 169003: 10365.7000   61.6765 CNS_WNV_D7_06     12</span></span>
<span><span class="co">#&gt; 169004:  9704.4700  -31.8532 CNS_WNV_D7_06     12</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="read-in-metadata">Read in metadata<a class="anchor" aria-label="anchor" href="#read-in-metadata"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Read in metadata  </span></span>
<span></span>
<span><span class="va">meta.dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">MetaDirectory</span>, <span class="st">"sample.details.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">meta.dat</span></span>
<span><span class="co">#&gt;              Filename     Sample  Group  Batch Cells per sample</span></span>
<span><span class="co">#&gt;                &lt;char&gt;     &lt;char&gt; &lt;char&gt; &lt;char&gt;            &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:   CNS_Mock_01.csv 01_Mock_01   Mock      A           420000</span></span>
<span><span class="co">#&gt;  2:   CNS_Mock_02.csv 02_Mock_02   Mock      B           240000</span></span>
<span><span class="co">#&gt;  3:   CNS_Mock_03.csv 03_Mock_03   Mock      B           256000</span></span>
<span><span class="co">#&gt;  4:   CNS_Mock_04.csv 04_Mock_04   Mock      A           252000</span></span>
<span><span class="co">#&gt;  5:   CNS_Mock_05.csv 05_Mock_05   Mock      A           345000</span></span>
<span><span class="co">#&gt;  6:   CNS_Mock_06.csv 06_Mock_06   Mock      B           702000</span></span>
<span><span class="co">#&gt;  7: CNS_WNV_D7_01.csv  07_WNV_01    WNV      A          5070000</span></span>
<span><span class="co">#&gt;  8: CNS_WNV_D7_02.csv  08_WNV_02    WNV      B          2940000</span></span>
<span><span class="co">#&gt;  9: CNS_WNV_D7_03.csv  09_WNV_03    WNV      A          2120000</span></span>
<span><span class="co">#&gt; 10: CNS_WNV_D7_04.csv  10_WNV_04    WNV      A          4320000</span></span>
<span><span class="co">#&gt; 11: CNS_WNV_D7_05.csv  11_WNV_05    WNV      B          4080000</span></span>
<span><span class="co">#&gt; 12: CNS_WNV_D7_06.csv  12_WNV_06    WNV      A          1830000</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="data-transformation">3. Data transformation<a class="anchor" aria-label="anchor" href="#data-transformation"></a>
</h2>
<p>Before we perform clustering etc, we need to meaningfully transform
the data. For more information on why this is necessary, please see <a href="https://immunedynamics.io/spectre/tutorials/data_transformation/data_transformation.html" class="external-link">this
page</a>.</p>
<p><strong>Note</strong>: <em>If you have imported CSV (channel value)
files exported from FlowJo, then no data transformations are required,
and you can skip all of the arcsinh transformation steps and proceed
straight to adding the metadata. More information on the FCS, CSV scale,
and CSV channel value file types can be found <a href="https://immunedynamics.io/spectre/tutorials/data_transformation/data_transformation.html" class="external-link">here</a>.</em></p>
<p>Create an output directory to store the transformation plots.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot.path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">OutputDirectory</span>, <span class="st">"Output 1 - transformed plots"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">plot.path</span><span class="op">)</span></span></code></pre></div>
<p>First, check the column names of the dataset.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cell.dat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;       [,1]      </span></span>
<span><span class="co">#&gt;  [1,] "NK11"    </span></span>
<span><span class="co">#&gt;  [2,] "CD3"     </span></span>
<span><span class="co">#&gt;  [3,] "CD45"    </span></span>
<span><span class="co">#&gt;  [4,] "Ly6G"    </span></span>
<span><span class="co">#&gt;  [5,] "CD11b"   </span></span>
<span><span class="co">#&gt;  [6,] "B220"    </span></span>
<span><span class="co">#&gt;  [7,] "CD8a"    </span></span>
<span><span class="co">#&gt;  [8,] "Ly6C"    </span></span>
<span><span class="co">#&gt;  [9,] "CD4"     </span></span>
<span><span class="co">#&gt; [10,] "FileName"</span></span>
<span><span class="co">#&gt; [11,] "FileNo"</span></span></code></pre></div>
<p>The columns we want to apply arcsinh transformation to are the
cellular columns – column 1 to column 9. We can specify those columns
using the code below.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Arcsinh transformation</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cell.dat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;       [,1]      </span></span>
<span><span class="co">#&gt;  [1,] "NK11"    </span></span>
<span><span class="co">#&gt;  [2,] "CD3"     </span></span>
<span><span class="co">#&gt;  [3,] "CD45"    </span></span>
<span><span class="co">#&gt;  [4,] "Ly6G"    </span></span>
<span><span class="co">#&gt;  [5,] "CD11b"   </span></span>
<span><span class="co">#&gt;  [6,] "B220"    </span></span>
<span><span class="co">#&gt;  [7,] "CD8a"    </span></span>
<span><span class="co">#&gt;  [8,] "Ly6C"    </span></span>
<span><span class="co">#&gt;  [9,] "CD4"     </span></span>
<span><span class="co">#&gt; [10,] "FileName"</span></span>
<span><span class="co">#&gt; [11,] "FileNo"</span></span>
<span></span>
<span><span class="va">to.asinh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cell.dat</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">to.asinh</span></span>
<span><span class="co">#&gt; [1] "NK11"  "CD3"   "CD45"  "Ly6G"  "CD11b" "B220"  "CD8a"  "Ly6C"  "CD4"</span></span></code></pre></div>
<p>Define the cofactor we will use for transformation. As a general
recommendation, we suggest using cofactor = 15 for CyTOF data, and
cofactor between 100 and 1000 for flow data (we suggest 500 as a
starting point). Here is a quick comparison figure showing how different
co-factors compare to bi-exponential transformations performed on an
LSR-II. For more detailed information on this choice, and for approaches
where different cofactors for different columns might be required, see
<a href="https://wiki.centenary.org.au/x/w8yACQ" class="external-link">this page</a>.</p>
<p><img src="images/simple_discovery/image2020-8-21_9-36-58.png" style="width:70.0%"></p>
<p><br></p>
<p>In this worked example we will use a cofactor of 500.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cofactor</span> <span class="op">&lt;-</span> <span class="fl">500</span></span></code></pre></div>
<p>You can also choose a column to use for plotting the transformed
result – ideally something that is expressed on a variety of cell types
in your dataset.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot.against</span> <span class="op">&lt;-</span> <span class="st">"Ly6C_asinh"</span></span></code></pre></div>
<p>Now we need to apply arcsinh transformation to the data in those
columns, using a specific co-factor.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell.dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do.asinh.html">do.asinh</a></span><span class="op">(</span><span class="va">cell.dat</span>, <span class="va">to.asinh</span>, cofactor <span class="op">=</span> <span class="va">cofactor</span><span class="op">)</span></span>
<span><span class="va">transformed.cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">to.asinh</span>, <span class="st">"_asinh"</span><span class="op">)</span></span></code></pre></div>
<p>We can then make some plots to see if the arcsinh transformation is
appropriate</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="va">transformed.cols</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/make.colour.plot.html">make.colour.plot</a></span><span class="op">(</span><span class="fu"><a href="../reference/do.subsample.html">do.subsample</a></span><span class="op">(</span><span class="va">cell.dat</span>, <span class="fl">20000</span><span class="op">)</span>, <span class="va">i</span>, <span class="va">plot.against</span>, path <span class="op">=</span> <span class="va">plot.path</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Check the plots and see if you are happy with the transformation. For
more detailed guidance, see <a href="https://immunedynamics.io/spectre/tutorials/data_transformation/data_transformation.html" class="external-link">this
page</a>. If happy, then proceed with analysis. Otherwise, go back to
the merging of the data.list (to create cell.dat) and try with another
co-factor.</p>
<p><img src="images/simple_discovery/image2020-8-20_13-34-47.png" style="width:70.0%"></p>
</div>
<div class="section level2">
<h2 id="add-metadata">4. Add metadata<a class="anchor" aria-label="anchor" href="#add-metadata"></a>
</h2>
<p>We also want to read in and attach some sample metadata, to aid with
our analysis.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Add metadata to data.table</span></span>
<span></span>
<span><span class="va">meta.dat</span></span>
<span><span class="co">#&gt;              Filename     Sample  Group  Batch Cells per sample</span></span>
<span><span class="co">#&gt;                &lt;char&gt;     &lt;char&gt; &lt;char&gt; &lt;char&gt;            &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:   CNS_Mock_01.csv 01_Mock_01   Mock      A           420000</span></span>
<span><span class="co">#&gt;  2:   CNS_Mock_02.csv 02_Mock_02   Mock      B           240000</span></span>
<span><span class="co">#&gt;  3:   CNS_Mock_03.csv 03_Mock_03   Mock      B           256000</span></span>
<span><span class="co">#&gt;  4:   CNS_Mock_04.csv 04_Mock_04   Mock      A           252000</span></span>
<span><span class="co">#&gt;  5:   CNS_Mock_05.csv 05_Mock_05   Mock      A           345000</span></span>
<span><span class="co">#&gt;  6:   CNS_Mock_06.csv 06_Mock_06   Mock      B           702000</span></span>
<span><span class="co">#&gt;  7: CNS_WNV_D7_01.csv  07_WNV_01    WNV      A          5070000</span></span>
<span><span class="co">#&gt;  8: CNS_WNV_D7_02.csv  08_WNV_02    WNV      B          2940000</span></span>
<span><span class="co">#&gt;  9: CNS_WNV_D7_03.csv  09_WNV_03    WNV      A          2120000</span></span>
<span><span class="co">#&gt; 10: CNS_WNV_D7_04.csv  10_WNV_04    WNV      A          4320000</span></span>
<span><span class="co">#&gt; 11: CNS_WNV_D7_05.csv  11_WNV_05    WNV      B          4080000</span></span>
<span><span class="co">#&gt; 12: CNS_WNV_D7_06.csv  12_WNV_06    WNV      A          1830000</span></span>
<span><span class="va">sample.info</span> <span class="op">&lt;-</span> <span class="va">meta.dat</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">sample.info</span></span>
<span><span class="co">#&gt;              Filename     Sample  Group  Batch</span></span>
<span><span class="co">#&gt;                &lt;char&gt;     &lt;char&gt; &lt;char&gt; &lt;char&gt;</span></span>
<span><span class="co">#&gt;  1:   CNS_Mock_01.csv 01_Mock_01   Mock      A</span></span>
<span><span class="co">#&gt;  2:   CNS_Mock_02.csv 02_Mock_02   Mock      B</span></span>
<span><span class="co">#&gt;  3:   CNS_Mock_03.csv 03_Mock_03   Mock      B</span></span>
<span><span class="co">#&gt;  4:   CNS_Mock_04.csv 04_Mock_04   Mock      A</span></span>
<span><span class="co">#&gt;  5:   CNS_Mock_05.csv 05_Mock_05   Mock      A</span></span>
<span><span class="co">#&gt;  6:   CNS_Mock_06.csv 06_Mock_06   Mock      B</span></span>
<span><span class="co">#&gt;  7: CNS_WNV_D7_01.csv  07_WNV_01    WNV      A</span></span>
<span><span class="co">#&gt;  8: CNS_WNV_D7_02.csv  08_WNV_02    WNV      B</span></span>
<span><span class="co">#&gt;  9: CNS_WNV_D7_03.csv  09_WNV_03    WNV      A</span></span>
<span><span class="co">#&gt; 10: CNS_WNV_D7_04.csv  10_WNV_04    WNV      A</span></span>
<span><span class="co">#&gt; 11: CNS_WNV_D7_05.csv  11_WNV_05    WNV      B</span></span>
<span><span class="co">#&gt; 12: CNS_WNV_D7_06.csv  12_WNV_06    WNV      A</span></span></code></pre></div>
<p>Once we have the metadata read into R, we will select only the
columns we want to add to our dataset. In this example we only want to
include use first four columns (Filename, Sample, Group, and Batch).
‘Filename’ will be used to for matching between cell.dat and meta.dat,
and the other three columns will be the information that gets added to
cell.dat</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta.dat</span></span>
<span><span class="co">#&gt;              Filename     Sample  Group  Batch Cells per sample</span></span>
<span><span class="co">#&gt;                &lt;char&gt;     &lt;char&gt; &lt;char&gt; &lt;char&gt;            &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:   CNS_Mock_01.csv 01_Mock_01   Mock      A           420000</span></span>
<span><span class="co">#&gt;  2:   CNS_Mock_02.csv 02_Mock_02   Mock      B           240000</span></span>
<span><span class="co">#&gt;  3:   CNS_Mock_03.csv 03_Mock_03   Mock      B           256000</span></span>
<span><span class="co">#&gt;  4:   CNS_Mock_04.csv 04_Mock_04   Mock      A           252000</span></span>
<span><span class="co">#&gt;  5:   CNS_Mock_05.csv 05_Mock_05   Mock      A           345000</span></span>
<span><span class="co">#&gt;  6:   CNS_Mock_06.csv 06_Mock_06   Mock      B           702000</span></span>
<span><span class="co">#&gt;  7: CNS_WNV_D7_01.csv  07_WNV_01    WNV      A          5070000</span></span>
<span><span class="co">#&gt;  8: CNS_WNV_D7_02.csv  08_WNV_02    WNV      B          2940000</span></span>
<span><span class="co">#&gt;  9: CNS_WNV_D7_03.csv  09_WNV_03    WNV      A          2120000</span></span>
<span><span class="co">#&gt; 10: CNS_WNV_D7_04.csv  10_WNV_04    WNV      A          4320000</span></span>
<span><span class="co">#&gt; 11: CNS_WNV_D7_05.csv  11_WNV_05    WNV      B          4080000</span></span>
<span><span class="co">#&gt; 12: CNS_WNV_D7_06.csv  12_WNV_06    WNV      A          1830000</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">meta.dat</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">5</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">counts</span></span>
<span><span class="co">#&gt;         Sample Cells per sample</span></span>
<span><span class="co">#&gt;         &lt;char&gt;            &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1: 01_Mock_01           420000</span></span>
<span><span class="co">#&gt;  2: 02_Mock_02           240000</span></span>
<span><span class="co">#&gt;  3: 03_Mock_03           256000</span></span>
<span><span class="co">#&gt;  4: 04_Mock_04           252000</span></span>
<span><span class="co">#&gt;  5: 05_Mock_05           345000</span></span>
<span><span class="co">#&gt;  6: 06_Mock_06           702000</span></span>
<span><span class="co">#&gt;  7:  07_WNV_01          5070000</span></span>
<span><span class="co">#&gt;  8:  08_WNV_02          2940000</span></span>
<span><span class="co">#&gt;  9:  09_WNV_03          2120000</span></span>
<span><span class="co">#&gt; 10:  10_WNV_04          4320000</span></span>
<span><span class="co">#&gt; 11:  11_WNV_05          4080000</span></span>
<span><span class="co">#&gt; 12:  12_WNV_06          1830000</span></span></code></pre></div>
<p>Now we can add this information to cell.dat. Essentially, the file
names are listed in the metadata table, and we can use that to add any
listed metadata in the table to the corresponding files in data.list. We
can thrn review the data to ensure the metadata has been correctly
embedded.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell.dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do.add.cols.html">do.add.cols</a></span><span class="op">(</span><span class="va">cell.dat</span>, <span class="st">"FileName"</span>, <span class="va">sample.info</span>, <span class="st">"Filename"</span>, rmv.ext <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">cell.dat</span></span>
<span><span class="co">#&gt;              NK11        CD3     CD45       Ly6G    CD11b      B220      CD8a</span></span>
<span><span class="co">#&gt;             &lt;num&gt;      &lt;num&gt;    &lt;num&gt;      &lt;num&gt;    &lt;num&gt;     &lt;num&gt;     &lt;num&gt;</span></span>
<span><span class="co">#&gt;      1:   42.3719  40.098700  6885.08  -344.7830 14787.30  -40.2399   83.7175</span></span>
<span><span class="co">#&gt;      2:   42.9586 119.014000  1780.29  -429.6650  5665.73   86.6673   34.7219</span></span>
<span><span class="co">#&gt;      3:   59.2366 206.238000 10248.30 -1603.8400 19894.30  427.8310  285.8800</span></span>
<span><span class="co">#&gt;      4:  364.9480  -0.233878  3740.04  -815.9800  9509.43  182.4200  333.6050</span></span>
<span><span class="co">#&gt;      5:  440.2470  40.035200  9191.38    40.5055  5745.82 -211.6940  149.2200</span></span>
<span><span class="co">#&gt;     ---                                                                      </span></span>
<span><span class="co">#&gt; 169000:  910.8890  72.856100 31466.20  -316.5570 28467.80   -7.7972 -271.8040</span></span>
<span><span class="co">#&gt; 169001:  -10.2642  64.188700 45188.00  -540.5140 22734.00  202.4110 -936.4920</span></span>
<span><span class="co">#&gt; 169002: -184.2910  -9.445650 11842.60   -97.9383 17237.00  123.4760 -219.9320</span></span>
<span><span class="co">#&gt; 169003:  248.3860 229.986000 32288.20  -681.1630 19255.80 -656.0540 -201.5880</span></span>
<span><span class="co">#&gt; 169004:  738.9810  95.470300 46185.10 -1004.6000 22957.80 -661.6280   72.3356</span></span>
<span><span class="co">#&gt;               Ly6C       CD4      FileName FileNo  NK11_asinh    CD3_asinh</span></span>
<span><span class="co">#&gt;              &lt;num&gt;     &lt;num&gt;        &lt;fctr&gt;  &lt;int&gt;       &lt;num&gt;        &lt;num&gt;</span></span>
<span><span class="co">#&gt;      1:   958.7000  711.0720   CNS_Mock_01      1  0.08464269  0.080111681</span></span>
<span><span class="co">#&gt;      2:   448.2590  307.2720   CNS_Mock_01      1  0.08581185  0.235835773</span></span>
<span><span class="co">#&gt;      3:  1008.8300  707.0940   CNS_Mock_01      1  0.11819779  0.401593928</span></span>
<span><span class="co">#&gt;      4:   440.0710  249.7840   CNS_Mock_01      1  0.67698633 -0.000467756</span></span>
<span><span class="co">#&gt;      5:    87.4815  867.5700   CNS_Mock_01      1  0.79429776  0.079985087</span></span>
<span><span class="co">#&gt;     ---                                                                   </span></span>
<span><span class="co">#&gt; 169000: 12023.7000 1103.0500 CNS_WNV_D7_06     12  1.36096843  0.145201437</span></span>
<span><span class="co">#&gt; 169001:  4188.3300  315.9400 CNS_WNV_D7_06     12 -0.02052696  0.128027364</span></span>
<span><span class="co">#&gt; 169002:  8923.4000 -453.4640 CNS_WNV_D7_06     12 -0.36070893 -0.018890177</span></span>
<span><span class="co">#&gt; 169003: 10365.7000   61.6765 CNS_WNV_D7_06     12  0.47832275  0.445126321</span></span>
<span><span class="co">#&gt; 169004:  9704.4700  -31.8532 CNS_WNV_D7_06     12  1.18247624  0.189799003</span></span>
<span><span class="co">#&gt;         CD45_asinh  Ly6G_asinh CD11b_asinh  B220_asinh  CD8a_asinh Ly6C_asinh</span></span>
<span><span class="co">#&gt;              &lt;num&gt;       &lt;num&gt;       &lt;num&gt;       &lt;num&gt;       &lt;num&gt;      &lt;num&gt;</span></span>
<span><span class="co">#&gt;      1:   3.316967 -0.64409775    4.080349 -0.08039317  0.16666238  1.4060734</span></span>
<span><span class="co">#&gt;      2:   1.982231 -0.77832998    3.122671  0.17247816  0.06938811  0.8062765</span></span>
<span><span class="co">#&gt;      3:   3.714001 -1.88215184    4.376885  0.77554551  0.54445897  1.4515055</span></span>
<span><span class="co">#&gt;      4:   2.709829 -1.26580137    3.639269  0.35719570  0.62559714  0.7940335</span></span>
<span><span class="co">#&gt;      5:   3.605299  0.08092265    3.136655 -0.41166199  0.29417852  0.1740824</span></span>
<span><span class="co">#&gt;     ---                                                                      </span></span>
<span><span class="co">#&gt; 169000:   4.835271 -0.59701179    4.735145 -0.01559377 -0.51987213  3.8736061</span></span>
<span><span class="co">#&gt; 169001:   5.197157 -0.93752450    4.510277  0.39450870 -1.38534526  2.8221404</span></span>
<span><span class="co">#&gt; 169002:   3.858443 -0.19464519    4.233563  0.24450841 -0.42678897  3.5757554</span></span>
<span><span class="co">#&gt; 169003:   4.861056 -1.11588765    4.344275 -1.08581190 -0.39298254  3.7253778</span></span>
<span><span class="co">#&gt; 169004:   5.218981 -1.44774229    4.520071 -1.09255123  0.14417124  3.6595440</span></span>
<span><span class="co">#&gt;           CD4_asinh     Sample  Group  Batch</span></span>
<span><span class="co">#&gt;               &lt;num&gt;     &lt;char&gt; &lt;char&gt; &lt;char&gt;</span></span>
<span><span class="co">#&gt;      1:  1.15078593 01_Mock_01   Mock      A</span></span>
<span><span class="co">#&gt;      2:  0.58125620 01_Mock_01   Mock      A</span></span>
<span><span class="co">#&gt;      3:  1.14620108 01_Mock_01   Mock      A</span></span>
<span><span class="co">#&gt;      4:  0.48082540 01_Mock_01   Mock      A</span></span>
<span><span class="co">#&gt;      5:  1.31850146 01_Mock_01   Mock      A</span></span>
<span><span class="co">#&gt;     ---                                     </span></span>
<span><span class="co">#&gt; 169000:  1.53218180  12_WNV_06    WNV      A</span></span>
<span><span class="co">#&gt; 169001:  0.59596889  12_WNV_06    WNV      A</span></span>
<span><span class="co">#&gt; 169002: -0.81400762  12_WNV_06    WNV      A</span></span>
<span><span class="co">#&gt; 169003:  0.12304230  12_WNV_06    WNV      A</span></span>
<span><span class="co">#&gt; 169004: -0.06366339  12_WNV_06    WNV      A</span></span></code></pre></div>
<p>Check the column names and specify columns that represent cellular
features (in this case, the arcsinh transformed data, defined by
“markername_asinh”).</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Columns</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cell.dat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;       [,1]         </span></span>
<span><span class="co">#&gt;  [1,] "NK11"       </span></span>
<span><span class="co">#&gt;  [2,] "CD3"        </span></span>
<span><span class="co">#&gt;  [3,] "CD45"       </span></span>
<span><span class="co">#&gt;  [4,] "Ly6G"       </span></span>
<span><span class="co">#&gt;  [5,] "CD11b"      </span></span>
<span><span class="co">#&gt;  [6,] "B220"       </span></span>
<span><span class="co">#&gt;  [7,] "CD8a"       </span></span>
<span><span class="co">#&gt;  [8,] "Ly6C"       </span></span>
<span><span class="co">#&gt;  [9,] "CD4"        </span></span>
<span><span class="co">#&gt; [10,] "FileName"   </span></span>
<span><span class="co">#&gt; [11,] "FileNo"     </span></span>
<span><span class="co">#&gt; [12,] "NK11_asinh" </span></span>
<span><span class="co">#&gt; [13,] "CD3_asinh"  </span></span>
<span><span class="co">#&gt; [14,] "CD45_asinh" </span></span>
<span><span class="co">#&gt; [15,] "Ly6G_asinh" </span></span>
<span><span class="co">#&gt; [16,] "CD11b_asinh"</span></span>
<span><span class="co">#&gt; [17,] "B220_asinh" </span></span>
<span><span class="co">#&gt; [18,] "CD8a_asinh" </span></span>
<span><span class="co">#&gt; [19,] "Ly6C_asinh" </span></span>
<span><span class="co">#&gt; [20,] "CD4_asinh"  </span></span>
<span><span class="co">#&gt; [21,] "Sample"     </span></span>
<span><span class="co">#&gt; [22,] "Group"      </span></span>
<span><span class="co">#&gt; [23,] "Batch"</span></span>
<span><span class="va">cellular.cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cell.dat</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">12</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">cellular.cols</span><span class="op">)</span></span>
<span><span class="co">#&gt;       [,1]         </span></span>
<span><span class="co">#&gt;  [1,] "NK11_asinh" </span></span>
<span><span class="co">#&gt;  [2,] "CD3_asinh"  </span></span>
<span><span class="co">#&gt;  [3,] "CD45_asinh" </span></span>
<span><span class="co">#&gt;  [4,] "Ly6G_asinh" </span></span>
<span><span class="co">#&gt;  [5,] "CD11b_asinh"</span></span>
<span><span class="co">#&gt;  [6,] "B220_asinh" </span></span>
<span><span class="co">#&gt;  [7,] "CD8a_asinh" </span></span>
<span><span class="co">#&gt;  [8,] "Ly6C_asinh" </span></span>
<span><span class="co">#&gt;  [9,] "CD4_asinh"</span></span></code></pre></div>
<p>Additionally, specify the columns that will be used to generate
cluster and tSNE/UMAP results. Columns that are not specified here will
still be analysed, but won’t contributed to the generation of clusters.
There are a couple of strategies to take here: use all cellular columns
for clustering to looks for all possible cell types/states, or use only
stably expressed markers to cluster stable phenotypes, which can then be
examined for changes in more dynamic markers. For more guidance, see our
<a href="https://immunedynamics.io/spectre/tutorials" class="external-link">tutorials
page</a>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cell.dat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;       [,1]         </span></span>
<span><span class="co">#&gt;  [1,] "NK11"       </span></span>
<span><span class="co">#&gt;  [2,] "CD3"        </span></span>
<span><span class="co">#&gt;  [3,] "CD45"       </span></span>
<span><span class="co">#&gt;  [4,] "Ly6G"       </span></span>
<span><span class="co">#&gt;  [5,] "CD11b"      </span></span>
<span><span class="co">#&gt;  [6,] "B220"       </span></span>
<span><span class="co">#&gt;  [7,] "CD8a"       </span></span>
<span><span class="co">#&gt;  [8,] "Ly6C"       </span></span>
<span><span class="co">#&gt;  [9,] "CD4"        </span></span>
<span><span class="co">#&gt; [10,] "FileName"   </span></span>
<span><span class="co">#&gt; [11,] "FileNo"     </span></span>
<span><span class="co">#&gt; [12,] "NK11_asinh" </span></span>
<span><span class="co">#&gt; [13,] "CD3_asinh"  </span></span>
<span><span class="co">#&gt; [14,] "CD45_asinh" </span></span>
<span><span class="co">#&gt; [15,] "Ly6G_asinh" </span></span>
<span><span class="co">#&gt; [16,] "CD11b_asinh"</span></span>
<span><span class="co">#&gt; [17,] "B220_asinh" </span></span>
<span><span class="co">#&gt; [18,] "CD8a_asinh" </span></span>
<span><span class="co">#&gt; [19,] "Ly6C_asinh" </span></span>
<span><span class="co">#&gt; [20,] "CD4_asinh"  </span></span>
<span><span class="co">#&gt; [21,] "Sample"     </span></span>
<span><span class="co">#&gt; [22,] "Group"      </span></span>
<span><span class="co">#&gt; [23,] "Batch"</span></span>
<span><span class="va">cluster.cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cell.dat</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">12</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">cluster.cols</span><span class="op">)</span></span>
<span><span class="co">#&gt;       [,1]         </span></span>
<span><span class="co">#&gt;  [1,] "NK11_asinh" </span></span>
<span><span class="co">#&gt;  [2,] "CD3_asinh"  </span></span>
<span><span class="co">#&gt;  [3,] "CD45_asinh" </span></span>
<span><span class="co">#&gt;  [4,] "Ly6G_asinh" </span></span>
<span><span class="co">#&gt;  [5,] "CD11b_asinh"</span></span>
<span><span class="co">#&gt;  [6,] "B220_asinh" </span></span>
<span><span class="co">#&gt;  [7,] "CD8a_asinh" </span></span>
<span><span class="co">#&gt;  [8,] "Ly6C_asinh" </span></span>
<span><span class="co">#&gt;  [9,] "CD4_asinh"</span></span></code></pre></div>
<p>Specify sample, group, and batch columns.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exp.name</span> <span class="op">&lt;-</span> <span class="st">"CNS experiment"</span></span>
<span><span class="va">sample.col</span> <span class="op">&lt;-</span> <span class="st">"Sample"</span></span>
<span><span class="va">group.col</span> <span class="op">&lt;-</span> <span class="st">"Group"</span></span>
<span><span class="va">batch.col</span> <span class="op">&lt;-</span> <span class="st">"Batch"</span></span></code></pre></div>
<p>Additionally, we want to specify the downsample targets for
dimensionality reduction. This influences how many cells will be shown
on a tSNE/UMAP plot, and we are specifying the number of cells per group
to downsample to. Check for the number of cells (rows) in each
group:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Subsample targets per group</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">cell.dat</span><span class="op">[[</span><span class="va">group.col</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="co"># Check number of cells per sample.</span></span>
<span><span class="co">#&gt;   Var1   Freq</span></span>
<span><span class="co">#&gt; 1 Mock  66992</span></span>
<span><span class="co">#&gt; 2  WNV 102012</span></span></code></pre></div>
<p>You can then specify the number to downsample to in each group. These
must be lower than the total number of cells in each group. In this
example we want 2000 cells from ‘mock’ and 20,000 cells from ‘WNV’, to
reflect the number of cells present in each group.</p>
<p>First, check the order that the groups appear in the dataset.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cell.dat</span><span class="op">[[</span><span class="va">group.col</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Mock" "WNV"</span></span></code></pre></div>
<p>Now you can specify the targets (in the order of
<code>unique(cell.dat[[group.col]])</code> above).</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sub.targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2000</span>, <span class="fl">20000</span><span class="op">)</span> <span class="co"># target subsample numbers from each group</span></span>
<span><span class="va">sub.targets</span></span>
<span><span class="co">#&gt; [1]  2000 20000</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="clustering-and-dimensionality-reduction">5. Clustering and Dimensionality reduction<a class="anchor" aria-label="anchor" href="#clustering-and-dimensionality-reduction"></a>
</h2>
<p>Create an output directory to store the clustering and dimensionality
reduction plots.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot.path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">OutputDirectory</span>, <span class="st">"Output 2 - clustering"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">plot.path</span><span class="op">)</span></span></code></pre></div>
<p>We can run clustering using the run.flowsom function. In this case we
can define the number of desired metaclusters manually, with the meta.k
argument (in this case we have chosen 8). This can be increased or
decreased as required. Typically, overclustering is preferred, as
multiple clusters that represent a single cellular population can always
be annotated as such. Subsequently, we can write the clustered dataset
to disk.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Clustering</span></span>
<span></span>
<span><span class="va">cell.dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run.flowsom.html">run.flowsom</a></span><span class="op">(</span><span class="va">cell.dat</span>, <span class="va">cluster.cols</span>, meta.k <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span><span class="va">cell.dat</span></span></code></pre></div>
<p>We can then run dimensionality reduction on a subset of the data,
allow us to visualise the data and resulting clusters. In this case we
have used run.umap, though other options are available, including
run.fitsne and run.tsne. As before, this subsampled dataset with DR
coordinates is saved to disk.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Dimensionality reduction</span></span>
<span></span>
<span><span class="va">cell.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do.subsample.html">do.subsample</a></span><span class="op">(</span><span class="va">cell.dat</span>, <span class="va">sub.targets</span>, <span class="va">group.col</span><span class="op">)</span></span>
<span><span class="va">cell.sub</span></span>
<span><span class="co">#&gt;             NK11       CD3     CD45      Ly6G    CD11b      B220      CD8a</span></span>
<span><span class="co">#&gt;            &lt;num&gt;     &lt;num&gt;    &lt;num&gt;     &lt;num&gt;    &lt;num&gt;     &lt;num&gt;     &lt;num&gt;</span></span>
<span><span class="co">#&gt;     1:  106.0150 235.13900 13574.70  -179.318 31385.80 -169.4210 -394.6730</span></span>
<span><span class="co">#&gt;     2:  132.7940  16.52790  9533.69  -673.468 22061.20  184.8820  282.5120</span></span>
<span><span class="co">#&gt;     3:  236.6740 143.33000  7805.94  -605.170 11658.20  658.9580  107.7280</span></span>
<span><span class="co">#&gt;     4:   14.8980 -10.63350  3628.48   468.071 12264.60 -178.1110 -194.3480</span></span>
<span><span class="co">#&gt;     5:  194.5450   3.62867  3192.66  -515.083 11939.20  -33.5803   42.3089</span></span>
<span><span class="co">#&gt;    ---                                                                    </span></span>
<span><span class="co">#&gt; 21996:  -16.4986  11.00010 24706.30  -263.460  9151.59 -144.2120 -499.9370</span></span>
<span><span class="co">#&gt; 21997:  119.8740  92.33010 43645.10 -1130.550 38278.10  -42.0887 -445.5530</span></span>
<span><span class="co">#&gt; 21998: 2884.3600  88.01560 88998.60 -4851.610  4889.69 4999.7900  -58.7703</span></span>
<span><span class="co">#&gt; 21999:  560.6400 151.51900 55440.80 -2633.550 23937.20  -95.7631  656.6650</span></span>
<span><span class="co">#&gt; 22000:  118.9280 331.26200 52654.70 -3161.200 25113.80 -938.5550  155.8260</span></span>
<span><span class="co">#&gt;             Ly6C       CD4      FileName FileNo  NK11_asinh    CD3_asinh</span></span>
<span><span class="co">#&gt;            &lt;num&gt;     &lt;num&gt;        &lt;fctr&gt;  &lt;int&gt;       &lt;num&gt;        &lt;num&gt;</span></span>
<span><span class="co">#&gt;     1:   380.647  509.4590   CNS_Mock_05      5  0.21047261  0.454470942</span></span>
<span><span class="co">#&gt;     2:   608.245 1095.4700   CNS_Mock_04      4  0.26256084  0.033049783</span></span>
<span><span class="co">#&gt;     3:   805.445 1937.4200   CNS_Mock_04      4  0.45724742  0.282872452</span></span>
<span><span class="co">#&gt;     4:   126.416  227.9000   CNS_Mock_04      4  0.02979159 -0.021265397</span></span>
<span><span class="co">#&gt;     5:   666.571  358.8580   CNS_Mock_02      2  0.37988669  0.007257276</span></span>
<span><span class="co">#&gt;    ---                                                                  </span></span>
<span><span class="co">#&gt; 21996: 22334.600   59.6839 CNS_WNV_D7_04     10 -0.03299121  0.021998426</span></span>
<span><span class="co">#&gt; 21997:  4490.860  727.2510 CNS_WNV_D7_06     12  0.23750870  0.183626518</span></span>
<span><span class="co">#&gt; 21998:  1175.860  273.2380 CNS_WNV_D7_01      7  2.45302657  0.175134535</span></span>
<span><span class="co">#&gt; 21999: 13667.400  334.3160 CNS_WNV_D7_04     10  0.96458591  0.298581704</span></span>
<span><span class="co">#&gt; 22000: 23457.800 2331.7400 CNS_WNV_D7_01      7  0.23566844  0.621694916</span></span>
<span><span class="co">#&gt;        CD45_asinh Ly6G_asinh CD11b_asinh  B220_asinh  CD8a_asinh Ly6C_asinh</span></span>
<span><span class="co">#&gt;             &lt;num&gt;      &lt;num&gt;       &lt;num&gt;       &lt;num&gt;       &lt;num&gt;      &lt;num&gt;</span></span>
<span><span class="co">#&gt;     1:   3.994841 -0.3513617    4.832713 -0.33267180 -0.72432725  0.7021579</span></span>
<span><span class="co">#&gt;     2:   3.641813 -1.1067473    4.480243  0.36181778  0.53860285  1.0264871</span></span>
<span><span class="co">#&gt;     3:   3.442203 -1.0225758    3.842764  1.08932754  0.21382293  1.2547409</span></span>
<span><span class="co">#&gt;     4:   2.679822  0.8354908    3.893427 -0.34908852 -0.37951948  0.2502130</span></span>
<span><span class="co">#&gt;     5:   2.553225 -0.9025441    3.866560 -0.06711021  0.08451714  1.0984975</span></span>
<span><span class="co">#&gt;    ---                                                                     </span></span>
<span><span class="co">#&gt; 21996:   4.593455 -0.5051593    3.600967 -0.28456776 -0.88128449  4.4925567</span></span>
<span><span class="co">#&gt; 21997:   5.162418 -1.5546568    5.031215 -0.08407830 -0.80224144  2.8914233</span></span>
<span><span class="co">#&gt; 21998:   5.874923 -2.9682497    2.976027  2.99818116 -0.11727162  1.5907079</span></span>
<span><span class="co">#&gt; 21999:   5.401630 -2.3635192    4.561837 -0.19037418  1.08655240  4.0016422</span></span>
<span><span class="co">#&gt; 22000:   5.350072 -2.5434425    4.609811 -1.38728687  0.30681557  4.5416110</span></span>
<span><span class="co">#&gt;        CD4_asinh     Sample  Group  Batch FlowSOM_cluster FlowSOM_metacluster</span></span>
<span><span class="co">#&gt;            &lt;num&gt;     &lt;char&gt; &lt;char&gt; &lt;char&gt;          &lt;fctr&gt;              &lt;fctr&gt;</span></span>
<span><span class="co">#&gt;     1: 0.8946876 05_Mock_05   Mock      A             190                   7</span></span>
<span><span class="co">#&gt;     2: 1.5259050 04_Mock_04   Mock      A             117                   7</span></span>
<span><span class="co">#&gt;     3: 2.0639011 04_Mock_04   Mock      A              60                   7</span></span>
<span><span class="co">#&gt;     4: 0.4413331 04_Mock_04   Mock      A             184                   7</span></span>
<span><span class="co">#&gt;     5: 0.6671197 02_Mock_02   Mock      B             147                   7</span></span>
<span><span class="co">#&gt;    ---                                                                       </span></span>
<span><span class="co">#&gt; 21996: 0.1190861  10_WNV_04    WNV      A             108                   2</span></span>
<span><span class="co">#&gt; 21997: 1.1692576  12_WNV_06    WNV      A              50                   2</span></span>
<span><span class="co">#&gt; 21998: 0.5223904  07_WNV_01    WNV      A              15                   1</span></span>
<span><span class="co">#&gt; 21999: 0.6267796  10_WNV_04    WNV      A              37                   2</span></span>
<span><span class="co">#&gt; 22000: 2.2442111  07_WNV_01    WNV      A              11                   2</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">cell.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run.umap.html">run.umap</a></span><span class="op">(</span><span class="va">cell.sub</span>, <span class="va">cluster.cols</span><span class="op">)</span></span>
<span><span class="va">cell.sub</span></span></code></pre></div>
<p>We can visualise the DR data to asses which clusters represent
cellular populations.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### DR plots</span></span>
<span></span>
<span><span class="fu"><a href="../reference/make.colour.plot.html">make.colour.plot</a></span><span class="op">(</span></span>
<span>    <span class="va">cell.sub</span>, </span>
<span>    <span class="st">"UMAP_X"</span>, </span>
<span>    <span class="st">"UMAP_Y"</span>, </span>
<span>    <span class="st">"FlowSOM_metacluster"</span>, </span>
<span>    col.type <span class="op">=</span> <span class="st">'factor'</span>, </span>
<span>    add.label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    path <span class="op">=</span> <span class="va">plot.path</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="simple_discovery_files/figure-html/unnamed-chunk-32-1.png" width="672"></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make.multi.plot.html">make.multi.plot</a></span><span class="op">(</span><span class="va">cell.sub</span>, <span class="st">"UMAP_X"</span>, <span class="st">"UMAP_Y"</span>, <span class="va">cellular.cols</span>, path <span class="op">=</span> <span class="va">plot.path</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/simple_discovery/Multi%20plot%20-%20plotted%20on%20UMAP_X%20by%20UMAP_Yx.png" style="width:70.0%"></p>
<p>We can also generate some multi plots to compare between experimental
groups or batches.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make.multi.plot.html">make.multi.plot</a></span><span class="op">(</span></span>
<span>    <span class="va">cell.sub</span>, <span class="st">"UMAP_X"</span>, <span class="st">"UMAP_Y"</span>, <span class="st">"FlowSOM_metacluster"</span>, </span>
<span>    <span class="va">group.col</span>, col.type <span class="op">=</span> <span class="st">'factor'</span>, path <span class="op">=</span> <span class="va">plot.path</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="images/simple_discovery/Multi%20plot%20divided%20by%20Group%20-%20plotted%20on%20UMAP_X%20by%20UMAP_Y.png" style="width:70.0%"></p>
<p>We can also produce expression heatmaps to help guide our
interpretation of cluster identities.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Expression heatmap</span></span>
<span></span>
<span><span class="va">exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do.aggregate.html">do.aggregate</a></span><span class="op">(</span><span class="va">cell.dat</span>, <span class="va">cellular.cols</span>, by <span class="op">=</span> <span class="st">"FlowSOM_metacluster"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/make.pheatmap.html">make.pheatmap</a></span><span class="op">(</span><span class="va">exp</span>, <span class="st">"FlowSOM_metacluster"</span>, <span class="va">cellular.cols</span>, path <span class="op">=</span> <span class="va">plot.path</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/simple_discovery/Pheatmap%20by%20FlowSOM_metacluster.png" style="width:50.0%"></p>
</div>
<div class="section level2">
<h2 id="annotate-clusters">6. Annotate clusters<a class="anchor" aria-label="anchor" href="#annotate-clusters"></a>
</h2>
<p>Review the cluster labels and marker expression patterns, so you can
annotate the clusters. This annotation is optional, as all subsequent
steps can be performed on the ‘clusters’ instead of the ‘populations’.
Here we can create a list of population names, and then specify which
clusters make up that population (e.g. CD4 T cells are contained within
cluster ‘3’).</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot.path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">OutputDirectory</span>, <span class="st">"Output 3 - annotation"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">plot.path</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### Annotate</span></span>
<span></span>
<span><span class="va">annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"CD4 T cells"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span>,</span>
<span>                <span class="st">"CD8 T cells"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>                <span class="st">"NK cells"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>                <span class="st">"Neutrophils"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span><span class="op">)</span>,</span>
<span>                <span class="st">"Infil Macrophages"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span>,</span>
<span>                <span class="st">"Microglia"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">6</span>,<span class="fl">7</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Once the annotation list is created, we can switch the list into a
table format to annotate our data.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do.list.switch.html">do.list.switch</a></span><span class="op">(</span><span class="va">annots</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">annots</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Values"</span>, <span class="st">"Population"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/setorder.html" class="external-link">setorderv</a></span><span class="op">(</span><span class="va">annots</span>, <span class="st">'Values'</span><span class="op">)</span></span>
<span><span class="va">annots</span></span>
<span><span class="co">#&gt;    Values        Population</span></span>
<span><span class="co">#&gt;     &lt;num&gt;            &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:      1          NK cells</span></span>
<span><span class="co">#&gt; 2:      2       CD8 T cells</span></span>
<span><span class="co">#&gt; 3:      3       CD4 T cells</span></span>
<span><span class="co">#&gt; 4:      4 Infil Macrophages</span></span>
<span><span class="co">#&gt; 5:      5         Microglia</span></span>
<span><span class="co">#&gt; 6:      6         Microglia</span></span>
<span><span class="co">#&gt; 7:      7         Microglia</span></span>
<span><span class="co">#&gt; 8:      8       Neutrophils</span></span></code></pre></div>
<p>Using the do.add.cols function, we can add the population names to
the corresponding clusters.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### Add annotations</span></span>
<span></span>
<span><span class="va">cell.dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do.add.cols.html">do.add.cols</a></span><span class="op">(</span><span class="va">cell.dat</span>, <span class="st">"FlowSOM_metacluster"</span>, <span class="va">annots</span>, <span class="st">"Values"</span><span class="op">)</span></span>
<span><span class="va">cell.dat</span></span>
<span><span class="co">#&gt;              NK11        CD3     CD45       Ly6G    CD11b      B220      CD8a</span></span>
<span><span class="co">#&gt;             &lt;num&gt;      &lt;num&gt;    &lt;num&gt;      &lt;num&gt;    &lt;num&gt;     &lt;num&gt;     &lt;num&gt;</span></span>
<span><span class="co">#&gt;      1:   42.3719  40.098700  6885.08  -344.7830 14787.30  -40.2399   83.7175</span></span>
<span><span class="co">#&gt;      2:   42.9586 119.014000  1780.29  -429.6650  5665.73   86.6673   34.7219</span></span>
<span><span class="co">#&gt;      3:   59.2366 206.238000 10248.30 -1603.8400 19894.30  427.8310  285.8800</span></span>
<span><span class="co">#&gt;      4:  364.9480  -0.233878  3740.04  -815.9800  9509.43  182.4200  333.6050</span></span>
<span><span class="co">#&gt;      5:  440.2470  40.035200  9191.38    40.5055  5745.82 -211.6940  149.2200</span></span>
<span><span class="co">#&gt;     ---                                                                      </span></span>
<span><span class="co">#&gt; 169000:  910.8890  72.856100 31466.20  -316.5570 28467.80   -7.7972 -271.8040</span></span>
<span><span class="co">#&gt; 169001:  -10.2642  64.188700 45188.00  -540.5140 22734.00  202.4110 -936.4920</span></span>
<span><span class="co">#&gt; 169002: -184.2910  -9.445650 11842.60   -97.9383 17237.00  123.4760 -219.9320</span></span>
<span><span class="co">#&gt; 169003:  248.3860 229.986000 32288.20  -681.1630 19255.80 -656.0540 -201.5880</span></span>
<span><span class="co">#&gt; 169004:  738.9810  95.470300 46185.10 -1004.6000 22957.80 -661.6280   72.3356</span></span>
<span><span class="co">#&gt;               Ly6C       CD4      FileName FileNo  NK11_asinh    CD3_asinh</span></span>
<span><span class="co">#&gt;              &lt;num&gt;     &lt;num&gt;        &lt;fctr&gt;  &lt;int&gt;       &lt;num&gt;        &lt;num&gt;</span></span>
<span><span class="co">#&gt;      1:   958.7000  711.0720   CNS_Mock_01      1  0.08464269  0.080111681</span></span>
<span><span class="co">#&gt;      2:   448.2590  307.2720   CNS_Mock_01      1  0.08581185  0.235835773</span></span>
<span><span class="co">#&gt;      3:  1008.8300  707.0940   CNS_Mock_01      1  0.11819779  0.401593928</span></span>
<span><span class="co">#&gt;      4:   440.0710  249.7840   CNS_Mock_01      1  0.67698633 -0.000467756</span></span>
<span><span class="co">#&gt;      5:    87.4815  867.5700   CNS_Mock_01      1  0.79429776  0.079985087</span></span>
<span><span class="co">#&gt;     ---                                                                   </span></span>
<span><span class="co">#&gt; 169000: 12023.7000 1103.0500 CNS_WNV_D7_06     12  1.36096843  0.145201437</span></span>
<span><span class="co">#&gt; 169001:  4188.3300  315.9400 CNS_WNV_D7_06     12 -0.02052696  0.128027364</span></span>
<span><span class="co">#&gt; 169002:  8923.4000 -453.4640 CNS_WNV_D7_06     12 -0.36070893 -0.018890177</span></span>
<span><span class="co">#&gt; 169003: 10365.7000   61.6765 CNS_WNV_D7_06     12  0.47832275  0.445126321</span></span>
<span><span class="co">#&gt; 169004:  9704.4700  -31.8532 CNS_WNV_D7_06     12  1.18247624  0.189799003</span></span>
<span><span class="co">#&gt;         CD45_asinh  Ly6G_asinh CD11b_asinh  B220_asinh  CD8a_asinh Ly6C_asinh</span></span>
<span><span class="co">#&gt;              &lt;num&gt;       &lt;num&gt;       &lt;num&gt;       &lt;num&gt;       &lt;num&gt;      &lt;num&gt;</span></span>
<span><span class="co">#&gt;      1:   3.316967 -0.64409775    4.080349 -0.08039317  0.16666238  1.4060734</span></span>
<span><span class="co">#&gt;      2:   1.982231 -0.77832998    3.122671  0.17247816  0.06938811  0.8062765</span></span>
<span><span class="co">#&gt;      3:   3.714001 -1.88215184    4.376885  0.77554551  0.54445897  1.4515055</span></span>
<span><span class="co">#&gt;      4:   2.709829 -1.26580137    3.639269  0.35719570  0.62559714  0.7940335</span></span>
<span><span class="co">#&gt;      5:   3.605299  0.08092265    3.136655 -0.41166199  0.29417852  0.1740824</span></span>
<span><span class="co">#&gt;     ---                                                                      </span></span>
<span><span class="co">#&gt; 169000:   4.835271 -0.59701179    4.735145 -0.01559377 -0.51987213  3.8736061</span></span>
<span><span class="co">#&gt; 169001:   5.197157 -0.93752450    4.510277  0.39450870 -1.38534526  2.8221404</span></span>
<span><span class="co">#&gt; 169002:   3.858443 -0.19464519    4.233563  0.24450841 -0.42678897  3.5757554</span></span>
<span><span class="co">#&gt; 169003:   4.861056 -1.11588765    4.344275 -1.08581190 -0.39298254  3.7253778</span></span>
<span><span class="co">#&gt; 169004:   5.218981 -1.44774229    4.520071 -1.09255123  0.14417124  3.6595440</span></span>
<span><span class="co">#&gt;           CD4_asinh     Sample  Group  Batch FlowSOM_cluster</span></span>
<span><span class="co">#&gt;               &lt;num&gt;     &lt;char&gt; &lt;char&gt; &lt;char&gt;          &lt;fctr&gt;</span></span>
<span><span class="co">#&gt;      1:  1.15078593 01_Mock_01   Mock      A             132</span></span>
<span><span class="co">#&gt;      2:  0.58125620 01_Mock_01   Mock      A              99</span></span>
<span><span class="co">#&gt;      3:  1.14620108 01_Mock_01   Mock      A              75</span></span>
<span><span class="co">#&gt;      4:  0.48082540 01_Mock_01   Mock      A             114</span></span>
<span><span class="co">#&gt;      5:  1.31850146 01_Mock_01   Mock      A             159</span></span>
<span><span class="co">#&gt;     ---                                                     </span></span>
<span><span class="co">#&gt; 169000:  1.53218180  12_WNV_06    WNV      A             138</span></span>
<span><span class="co">#&gt; 169001:  0.59596889  12_WNV_06    WNV      A              92</span></span>
<span><span class="co">#&gt; 169002: -0.81400762  12_WNV_06    WNV      A             164</span></span>
<span><span class="co">#&gt; 169003:  0.12304230  12_WNV_06    WNV      A             107</span></span>
<span><span class="co">#&gt; 169004: -0.06366339  12_WNV_06    WNV      A              51</span></span>
<span><span class="co">#&gt;         FlowSOM_metacluster  Population</span></span>
<span><span class="co">#&gt;                      &lt;fctr&gt;      &lt;char&gt;</span></span>
<span><span class="co">#&gt;      1:                   7   Microglia</span></span>
<span><span class="co">#&gt;      2:                   7   Microglia</span></span>
<span><span class="co">#&gt;      3:                   7   Microglia</span></span>
<span><span class="co">#&gt;      4:                   7   Microglia</span></span>
<span><span class="co">#&gt;      5:                   7   Microglia</span></span>
<span><span class="co">#&gt;     ---                                </span></span>
<span><span class="co">#&gt; 169000:                   2 CD8 T cells</span></span>
<span><span class="co">#&gt; 169001:                   2 CD8 T cells</span></span>
<span><span class="co">#&gt; 169002:                   2 CD8 T cells</span></span>
<span><span class="co">#&gt; 169003:                   2 CD8 T cells</span></span>
<span><span class="co">#&gt; 169004:                   2 CD8 T cells</span></span>
<span></span>
<span><span class="va">cell.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do.add.cols.html">do.add.cols</a></span><span class="op">(</span><span class="va">cell.sub</span>, <span class="st">"FlowSOM_metacluster"</span>, <span class="va">annots</span>, <span class="st">"Values"</span><span class="op">)</span></span>
<span><span class="va">cell.sub</span></span>
<span><span class="co">#&gt;             NK11       CD3     CD45      Ly6G    CD11b      B220      CD8a</span></span>
<span><span class="co">#&gt;            &lt;num&gt;     &lt;num&gt;    &lt;num&gt;     &lt;num&gt;    &lt;num&gt;     &lt;num&gt;     &lt;num&gt;</span></span>
<span><span class="co">#&gt;     1:  106.0150 235.13900 13574.70  -179.318 31385.80 -169.4210 -394.6730</span></span>
<span><span class="co">#&gt;     2:  132.7940  16.52790  9533.69  -673.468 22061.20  184.8820  282.5120</span></span>
<span><span class="co">#&gt;     3:  236.6740 143.33000  7805.94  -605.170 11658.20  658.9580  107.7280</span></span>
<span><span class="co">#&gt;     4:   14.8980 -10.63350  3628.48   468.071 12264.60 -178.1110 -194.3480</span></span>
<span><span class="co">#&gt;     5:  194.5450   3.62867  3192.66  -515.083 11939.20  -33.5803   42.3089</span></span>
<span><span class="co">#&gt;    ---                                                                    </span></span>
<span><span class="co">#&gt; 21996:  -16.4986  11.00010 24706.30  -263.460  9151.59 -144.2120 -499.9370</span></span>
<span><span class="co">#&gt; 21997:  119.8740  92.33010 43645.10 -1130.550 38278.10  -42.0887 -445.5530</span></span>
<span><span class="co">#&gt; 21998: 2884.3600  88.01560 88998.60 -4851.610  4889.69 4999.7900  -58.7703</span></span>
<span><span class="co">#&gt; 21999:  560.6400 151.51900 55440.80 -2633.550 23937.20  -95.7631  656.6650</span></span>
<span><span class="co">#&gt; 22000:  118.9280 331.26200 52654.70 -3161.200 25113.80 -938.5550  155.8260</span></span>
<span><span class="co">#&gt;             Ly6C       CD4      FileName FileNo  NK11_asinh    CD3_asinh</span></span>
<span><span class="co">#&gt;            &lt;num&gt;     &lt;num&gt;        &lt;fctr&gt;  &lt;int&gt;       &lt;num&gt;        &lt;num&gt;</span></span>
<span><span class="co">#&gt;     1:   380.647  509.4590   CNS_Mock_05      5  0.21047261  0.454470942</span></span>
<span><span class="co">#&gt;     2:   608.245 1095.4700   CNS_Mock_04      4  0.26256084  0.033049783</span></span>
<span><span class="co">#&gt;     3:   805.445 1937.4200   CNS_Mock_04      4  0.45724742  0.282872452</span></span>
<span><span class="co">#&gt;     4:   126.416  227.9000   CNS_Mock_04      4  0.02979159 -0.021265397</span></span>
<span><span class="co">#&gt;     5:   666.571  358.8580   CNS_Mock_02      2  0.37988669  0.007257276</span></span>
<span><span class="co">#&gt;    ---                                                                  </span></span>
<span><span class="co">#&gt; 21996: 22334.600   59.6839 CNS_WNV_D7_04     10 -0.03299121  0.021998426</span></span>
<span><span class="co">#&gt; 21997:  4490.860  727.2510 CNS_WNV_D7_06     12  0.23750870  0.183626518</span></span>
<span><span class="co">#&gt; 21998:  1175.860  273.2380 CNS_WNV_D7_01      7  2.45302657  0.175134535</span></span>
<span><span class="co">#&gt; 21999: 13667.400  334.3160 CNS_WNV_D7_04     10  0.96458591  0.298581704</span></span>
<span><span class="co">#&gt; 22000: 23457.800 2331.7400 CNS_WNV_D7_01      7  0.23566844  0.621694916</span></span>
<span><span class="co">#&gt;        CD45_asinh Ly6G_asinh CD11b_asinh  B220_asinh  CD8a_asinh Ly6C_asinh</span></span>
<span><span class="co">#&gt;             &lt;num&gt;      &lt;num&gt;       &lt;num&gt;       &lt;num&gt;       &lt;num&gt;      &lt;num&gt;</span></span>
<span><span class="co">#&gt;     1:   3.994841 -0.3513617    4.832713 -0.33267180 -0.72432725  0.7021579</span></span>
<span><span class="co">#&gt;     2:   3.641813 -1.1067473    4.480243  0.36181778  0.53860285  1.0264871</span></span>
<span><span class="co">#&gt;     3:   3.442203 -1.0225758    3.842764  1.08932754  0.21382293  1.2547409</span></span>
<span><span class="co">#&gt;     4:   2.679822  0.8354908    3.893427 -0.34908852 -0.37951948  0.2502130</span></span>
<span><span class="co">#&gt;     5:   2.553225 -0.9025441    3.866560 -0.06711021  0.08451714  1.0984975</span></span>
<span><span class="co">#&gt;    ---                                                                     </span></span>
<span><span class="co">#&gt; 21996:   4.593455 -0.5051593    3.600967 -0.28456776 -0.88128449  4.4925567</span></span>
<span><span class="co">#&gt; 21997:   5.162418 -1.5546568    5.031215 -0.08407830 -0.80224144  2.8914233</span></span>
<span><span class="co">#&gt; 21998:   5.874923 -2.9682497    2.976027  2.99818116 -0.11727162  1.5907079</span></span>
<span><span class="co">#&gt; 21999:   5.401630 -2.3635192    4.561837 -0.19037418  1.08655240  4.0016422</span></span>
<span><span class="co">#&gt; 22000:   5.350072 -2.5434425    4.609811 -1.38728687  0.30681557  4.5416110</span></span>
<span><span class="co">#&gt;        CD4_asinh     Sample  Group  Batch FlowSOM_cluster FlowSOM_metacluster</span></span>
<span><span class="co">#&gt;            &lt;num&gt;     &lt;char&gt; &lt;char&gt; &lt;char&gt;          &lt;fctr&gt;              &lt;fctr&gt;</span></span>
<span><span class="co">#&gt;     1: 0.8946876 05_Mock_05   Mock      A             190                   7</span></span>
<span><span class="co">#&gt;     2: 1.5259050 04_Mock_04   Mock      A             117                   7</span></span>
<span><span class="co">#&gt;     3: 2.0639011 04_Mock_04   Mock      A              60                   7</span></span>
<span><span class="co">#&gt;     4: 0.4413331 04_Mock_04   Mock      A             184                   7</span></span>
<span><span class="co">#&gt;     5: 0.6671197 02_Mock_02   Mock      B             147                   7</span></span>
<span><span class="co">#&gt;    ---                                                                       </span></span>
<span><span class="co">#&gt; 21996: 0.1190861  10_WNV_04    WNV      A             108                   2</span></span>
<span><span class="co">#&gt; 21997: 1.1692576  12_WNV_06    WNV      A              50                   2</span></span>
<span><span class="co">#&gt; 21998: 0.5223904  07_WNV_01    WNV      A              15                   1</span></span>
<span><span class="co">#&gt; 21999: 0.6267796  10_WNV_04    WNV      A              37                   2</span></span>
<span><span class="co">#&gt; 22000: 2.2442111  07_WNV_01    WNV      A              11                   2</span></span>
<span><span class="co">#&gt;            UMAP_X     UMAP_Y  Population</span></span>
<span><span class="co">#&gt;             &lt;num&gt;      &lt;num&gt;      &lt;char&gt;</span></span>
<span><span class="co">#&gt;     1: -3.1579463 -1.8178011   Microglia</span></span>
<span><span class="co">#&gt;     2: -3.4301853 -0.7046464   Microglia</span></span>
<span><span class="co">#&gt;     3: -3.5628729 -0.9590427   Microglia</span></span>
<span><span class="co">#&gt;     4: -4.1040456 -3.1769343   Microglia</span></span>
<span><span class="co">#&gt;     5: -4.4687803 -1.7181371   Microglia</span></span>
<span><span class="co">#&gt;    ---                                  </span></span>
<span><span class="co">#&gt; 21996:  1.7882223 -1.0667364 CD8 T cells</span></span>
<span><span class="co">#&gt; 21997:  1.7846371  1.1294917 CD8 T cells</span></span>
<span><span class="co">#&gt; 21998: -3.3775246 -8.0603996    NK cells</span></span>
<span><span class="co">#&gt; 21999:  0.3071737  3.3264553 CD8 T cells</span></span>
<span><span class="co">#&gt; 22000:  1.1479282  5.0892695 CD8 T cells</span></span></code></pre></div>
<p>Subsequently, we can visualise the population labels on a UMAP
plot.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make.colour.plot.html">make.colour.plot</a></span><span class="op">(</span><span class="va">cell.sub</span>, <span class="st">"UMAP_X"</span>, <span class="st">"UMAP_Y"</span>, <span class="st">"Population"</span>, col.type <span class="op">=</span> <span class="st">'factor'</span>, add.label <span class="op">=</span> <span class="cn">TRUE</span>, path <span class="op">=</span> <span class="va">plot.path</span><span class="op">)</span></span></code></pre></div>
<p><img src="simple_discovery_files/figure-html/unnamed-chunk-40-1.png" width="672"></p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make.multi.plot.html">make.multi.plot</a></span><span class="op">(</span><span class="va">cell.sub</span>, <span class="st">"UMAP_X"</span>, <span class="st">"UMAP_Y"</span>, <span class="st">"Population"</span>, <span class="va">group.col</span>, col.type <span class="op">=</span> <span class="st">'factor'</span>, path <span class="op">=</span> <span class="va">plot.path</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/simple_discovery/Multi%20plot%20divided%20by%20Group%20-%20plotted%20on%20UMAP_X%20by%20UMAP_Yxx.png" style="width:70.0%"></p>
<p>We can also generate an expression heatmap to summarise the
expression levels of each marker on our populations.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### Expression heatmap</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">exp</span><span class="op">)</span></span>
<span><span class="va">exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do.aggregate.html">do.aggregate</a></span><span class="op">(</span><span class="va">cell.dat</span>, <span class="va">cellular.cols</span>, by <span class="op">=</span> <span class="st">"Population"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/make.pheatmap.html">make.pheatmap</a></span><span class="op">(</span><span class="va">exp</span>, <span class="st">"Population"</span>, <span class="va">cellular.cols</span>, path <span class="op">=</span> <span class="va">plot.path</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/simple_discovery/image2021-1-13_16-46-33.png" style="width:50.0%"></p>
<p>Save the annotated data to disk, both the full dataset (cell.dat) and
the downsampled dataset (cell.sub).</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Write data</span></span>
<span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">OutputDirectory</span>, <span class="st">"Output 3 - annotation"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outdir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fwrite.html" class="external-link">fwrite</a></span><span class="op">(</span><span class="va">cell.dat</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"Annotated.data.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fwrite.html" class="external-link">fwrite</a></span><span class="op">(</span><span class="va">cell.sub</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"Annotated.data.DR.csv"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Save data as FCS files for use in FlowJo or other software.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Write FCS files</span></span>
<span></span>
<span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">OutputDirectory</span>, <span class="st">"Output 3 - annotation"</span>, <span class="st">"FCS files - all"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outdir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/write.files.html">write.files</a></span><span class="op">(</span><span class="va">cell.dat</span>,</span>
<span>            file.prefix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="va">exp.name</span><span class="op">)</span>,</span>
<span>            divide.by <span class="op">=</span> <span class="va">sample.col</span>,</span>
<span>            write.csv <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            write.fcs <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">OutputDirectory</span>, <span class="st">"Output 3 - annotation"</span>, <span class="st">"FCS files - DR"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outdir</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/write.files.html">write.files</a></span><span class="op">(</span><span class="va">cell.sub</span>,</span>
<span>            file.prefix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'DR-'</span>, <span class="va">exp.name</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            divide.by <span class="op">=</span> <span class="va">sample.col</span>,</span>
<span>            write.csv <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            write.fcs <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="summary-data-and-stats">7. Summary data and stats<a class="anchor" aria-label="anchor" href="#summary-data-and-stats"></a>
</h2>
<p>Here we can create ‘summary’ data for our experiment. This involves
calculating the percentage of each population in each sample, along with
the corresponding cell counts if the information is available. In
addition, we calculate the MFI for selected markers on each population
in each sample.</p>
<p>First, set the working directory, and select which columns we will
measure the MFI of. In this case, CD11b_asinh and Ly6C_asinh.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">OutputDirectory</span>, <span class="st">"Output 4 - summary data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### Setup</span></span>
<span></span>
<span><span class="va">variance.test</span> <span class="op">&lt;-</span> <span class="st">'kruskal.test'</span></span>
<span><span class="va">pairwise.test</span> <span class="op">&lt;-</span> <span class="st">"wilcox.test"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cell.dat</span><span class="op">[[</span><span class="va">group.col</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1]  </span></span>
<span><span class="co">#&gt; [1,] "Mock"</span></span>
<span><span class="co">#&gt; [2,] "WNV"</span></span>
<span></span>
<span><span class="va">comparisons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Mock"</span>, <span class="st">"WNV"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">comparisons</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] "Mock" "WNV"</span></span>
<span></span>
<span><span class="va">grp.order</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Mock"</span>, <span class="st">"WNV"</span><span class="op">)</span></span>
<span><span class="va">grp.order</span></span>
<span><span class="co">#&gt; [1] "Mock" "WNV"</span></span></code></pre></div>
<p>We can also specify which columns we wish to measure MFI levels
on.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### Select columns to measure MFI</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">cellular.cols</span><span class="op">)</span></span>
<span><span class="co">#&gt;       [,1]         </span></span>
<span><span class="co">#&gt;  [1,] "NK11_asinh" </span></span>
<span><span class="co">#&gt;  [2,] "CD3_asinh"  </span></span>
<span><span class="co">#&gt;  [3,] "CD45_asinh" </span></span>
<span><span class="co">#&gt;  [4,] "Ly6G_asinh" </span></span>
<span><span class="co">#&gt;  [5,] "CD11b_asinh"</span></span>
<span><span class="co">#&gt;  [6,] "B220_asinh" </span></span>
<span><span class="co">#&gt;  [7,] "CD8a_asinh" </span></span>
<span><span class="co">#&gt;  [8,] "Ly6C_asinh" </span></span>
<span><span class="co">#&gt;  [9,] "CD4_asinh"</span></span>
<span><span class="va">dyn.cols</span> <span class="op">&lt;-</span> <span class="va">cellular.cols</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">8</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">dyn.cols</span></span>
<span><span class="co">#&gt; [1] "CD11b_asinh" "Ly6C_asinh"</span></span></code></pre></div>
<div class="section level3">
<h3 id="create-summary-data">Create summary data<a class="anchor" aria-label="anchor" href="#create-summary-data"></a>
</h3>
<p>Use the new create.sumtable function to generate summary data – a
data.table of samples (rows) vs measurements (columns).</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### Create summary tables</span></span>
<span></span>
<span><span class="va">sum.dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create.sumtable.html">create.sumtable</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">cell.dat</span>, </span>
<span>                            sample.col <span class="op">=</span> <span class="va">sample.col</span>,</span>
<span>                            pop.col <span class="op">=</span> <span class="st">"Population"</span>,</span>
<span>                            use.cols <span class="op">=</span> <span class="va">dyn.cols</span>, </span>
<span>                            annot.cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">group.col</span>, <span class="va">batch.col</span><span class="op">)</span>, </span>
<span>                            counts <span class="op">=</span> <span class="va">counts</span><span class="op">)</span></span></code></pre></div>
<p>Once the summary data has been generated, we can review it and select
which columns to plot. In each case, the column names (i.e. name of each
summary measure) are structured as ‘MEASURE TYPE – POPULATION’. This
provides a useful structure, as we can use regular expression searches
to split the name into just the MEASURE TYPE or POPULATION segment.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### Review summary data</span></span>
<span></span>
<span><span class="va">sum.dat</span></span>
<span><span class="co">#&gt;         Sample  Group  Batch Cells per sample -- CD4 T cells</span></span>
<span><span class="co">#&gt;         &lt;fctr&gt; &lt;char&gt; &lt;char&gt;                           &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1: 01_Mock_01   Mock      A                       2789.5743</span></span>
<span><span class="co">#&gt;  2: 02_Mock_02   Mock      B                       1183.2631</span></span>
<span><span class="co">#&gt;  3: 03_Mock_03   Mock      B                       1257.8970</span></span>
<span><span class="co">#&gt;  4: 04_Mock_04   Mock      A                        487.9136</span></span>
<span><span class="co">#&gt;  5: 05_Mock_05   Mock      A                        568.7855</span></span>
<span><span class="co">#&gt;  6: 06_Mock_06   Mock      B                       2439.9918</span></span>
<span><span class="co">#&gt;  7:  07_WNV_01    WNV      A                     265221.7270</span></span>
<span><span class="co">#&gt;  8:  08_WNV_02    WNV      B                      58680.4657</span></span>
<span><span class="co">#&gt;  9:  09_WNV_03    WNV      A                      53930.8707</span></span>
<span><span class="co">#&gt; 10:  10_WNV_04    WNV      A                     109695.8730</span></span>
<span><span class="co">#&gt; 11:  11_WNV_05    WNV      B                     125018.2092</span></span>
<span><span class="co">#&gt; 12:  12_WNV_06    WNV      A                      61022.7454</span></span>
<span><span class="co">#&gt;     Cells per sample -- CD8 T cells Cells per sample -- Infil Macrophages</span></span>
<span><span class="co">#&gt;                               &lt;num&gt;                                 &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:                       11750.025                            10608.8357</span></span>
<span><span class="co">#&gt;  2:                        7084.009                             2382.0954</span></span>
<span><span class="co">#&gt;  3:                        7601.292                             2120.4549</span></span>
<span><span class="co">#&gt;  4:                        5426.191                              857.5452</span></span>
<span><span class="co">#&gt;  5:                       11691.702                             1579.9597</span></span>
<span><span class="co">#&gt;  6:                       46216.316                             8181.1490</span></span>
<span><span class="co">#&gt;  7:                     3177576.602                           238106.4067</span></span>
<span><span class="co">#&gt;  8:                     2046390.686                            66830.5304</span></span>
<span><span class="co">#&gt;  9:                     1480647.540                            72747.0467</span></span>
<span><span class="co">#&gt; 10:                     2908075.419                           130878.5243</span></span>
<span><span class="co">#&gt; 11:                     2692246.641                           143207.3339</span></span>
<span><span class="co">#&gt; 12:                     1229520.588                            68236.2968</span></span>
<span><span class="co">#&gt;     Cells per sample -- Microglia Cells per sample -- Neutrophils</span></span>
<span><span class="co">#&gt;                             &lt;num&gt;                           &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:                      380946.0                       11496.427</span></span>
<span><span class="co">#&gt;  2:                      220585.1                        6772.624</span></span>
<span><span class="co">#&gt;  3:                      234597.8                        7852.871</span></span>
<span><span class="co">#&gt;  4:                      238382.8                        5529.688</span></span>
<span><span class="co">#&gt;  5:                      317571.9                       10870.123</span></span>
<span><span class="co">#&gt;  6:                      587463.9                       52531.589</span></span>
<span><span class="co">#&gt;  7:                      634385.5                      119476.880</span></span>
<span><span class="co">#&gt;  8:                      366028.5                      144346.701</span></span>
<span><span class="co">#&gt;  9:                      255476.0                       83745.234</span></span>
<span><span class="co">#&gt; 10:                      486192.3                      348505.049</span></span>
<span><span class="co">#&gt; 11:                      493412.0                       73525.053</span></span>
<span><span class="co">#&gt; 12:                      245845.6                       35580.355</span></span>
<span><span class="co">#&gt;     Cells per sample -- NK cells Exp CD11b_asinh -- CD4 T cells</span></span>
<span><span class="co">#&gt;                            &lt;num&gt;                          &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:                     2409.178                      1.1640880</span></span>
<span><span class="co">#&gt;  2:                     1992.864                      1.0717827</span></span>
<span><span class="co">#&gt;  3:                     2569.704                      1.1071680</span></span>
<span><span class="co">#&gt;  4:                     1315.888                      0.9897486</span></span>
<span><span class="co">#&gt;  5:                     2717.531                      0.9672789</span></span>
<span><span class="co">#&gt;  6:                     5167.042                      0.9624663</span></span>
<span><span class="co">#&gt;  7:                   635232.869                      2.1897904</span></span>
<span><span class="co">#&gt;  8:                   257723.157                      2.1366929</span></span>
<span><span class="co">#&gt;  9:                   173453.341                      2.1843277</span></span>
<span><span class="co">#&gt; 10:                   336652.852                      1.9940074</span></span>
<span><span class="co">#&gt; 11:                   552590.732                      2.1698435</span></span>
<span><span class="co">#&gt; 12:                   189794.386                      2.1240310</span></span>
<span><span class="co">#&gt;     Exp CD11b_asinh -- CD8 T cells Exp CD11b_asinh -- Infil Macrophages</span></span>
<span><span class="co">#&gt;                              &lt;num&gt;                                &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:                       3.707953                            1.2200778</span></span>
<span><span class="co">#&gt;  2:                       3.932891                            1.2510496</span></span>
<span><span class="co">#&gt;  3:                       4.055855                            1.0506404</span></span>
<span><span class="co">#&gt;  4:                       3.980852                            1.0649411</span></span>
<span><span class="co">#&gt;  5:                       3.904824                            0.7475088</span></span>
<span><span class="co">#&gt;  6:                       3.533980                            0.9208460</span></span>
<span><span class="co">#&gt;  7:                       4.554693                            2.3009085</span></span>
<span><span class="co">#&gt;  8:                       4.485134                            1.9926056</span></span>
<span><span class="co">#&gt;  9:                       4.666380                            2.2236088</span></span>
<span><span class="co">#&gt; 10:                       4.425055                            2.0578310</span></span>
<span><span class="co">#&gt; 11:                       4.832076                            2.1922020</span></span>
<span><span class="co">#&gt; 12:                       4.685090                            2.2536754</span></span>
<span><span class="co">#&gt;     Exp CD11b_asinh -- Microglia Exp CD11b_asinh -- Neutrophils</span></span>
<span><span class="co">#&gt;                            &lt;num&gt;                          &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:                     3.703104                       4.424470</span></span>
<span><span class="co">#&gt;  2:                     3.684539                       4.319386</span></span>
<span><span class="co">#&gt;  3:                     3.726379                       4.491092</span></span>
<span><span class="co">#&gt;  4:                     3.785686                       4.261786</span></span>
<span><span class="co">#&gt;  5:                     3.747101                       4.369713</span></span>
<span><span class="co">#&gt;  6:                     3.722985                       4.529456</span></span>
<span><span class="co">#&gt;  7:                     3.652149                       5.626350</span></span>
<span><span class="co">#&gt;  8:                     3.584455                       5.782984</span></span>
<span><span class="co">#&gt;  9:                     3.803963                       5.638535</span></span>
<span><span class="co">#&gt; 10:                     3.681785                       5.818091</span></span>
<span><span class="co">#&gt; 11:                     3.663684                       5.417286</span></span>
<span><span class="co">#&gt; 12:                     3.795823                       5.384971</span></span>
<span><span class="co">#&gt;     Exp CD11b_asinh -- NK cells Exp Ly6C_asinh -- CD4 T cells</span></span>
<span><span class="co">#&gt;                           &lt;num&gt;                         &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:                    1.936982                     1.0181309</span></span>
<span><span class="co">#&gt;  2:                    1.834861                     1.3204695</span></span>
<span><span class="co">#&gt;  3:                    1.905189                     0.9998138</span></span>
<span><span class="co">#&gt;  4:                    1.685278                     0.4985209</span></span>
<span><span class="co">#&gt;  5:                    1.584877                     1.3079999</span></span>
<span><span class="co">#&gt;  6:                    1.555587                     0.9641838</span></span>
<span><span class="co">#&gt;  7:                    2.954674                     3.1288202</span></span>
<span><span class="co">#&gt;  8:                    2.972281                     2.5929207</span></span>
<span><span class="co">#&gt;  9:                    2.912120                     2.3828927</span></span>
<span><span class="co">#&gt; 10:                    2.870294                     2.6614759</span></span>
<span><span class="co">#&gt; 11:                    2.831210                     2.7495585</span></span>
<span><span class="co">#&gt; 12:                    2.729239                     2.2266242</span></span>
<span><span class="co">#&gt;     Exp Ly6C_asinh -- CD8 T cells Exp Ly6C_asinh -- Infil Macrophages</span></span>
<span><span class="co">#&gt;                             &lt;num&gt;                               &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:                      4.292275                            4.270251</span></span>
<span><span class="co">#&gt;  2:                      4.319977                            1.232527</span></span>
<span><span class="co">#&gt;  3:                      4.385810                            2.381495</span></span>
<span><span class="co">#&gt;  4:                      4.119795                            1.268299</span></span>
<span><span class="co">#&gt;  5:                      4.223415                            3.996084</span></span>
<span><span class="co">#&gt;  6:                      4.668365                            3.350074</span></span>
<span><span class="co">#&gt;  7:                      4.680534                            3.241140</span></span>
<span><span class="co">#&gt;  8:                      4.220517                            2.663728</span></span>
<span><span class="co">#&gt;  9:                      3.966745                            2.529144</span></span>
<span><span class="co">#&gt; 10:                      4.367482                            2.649219</span></span>
<span><span class="co">#&gt; 11:                      4.186924                            2.915013</span></span>
<span><span class="co">#&gt; 12:                      3.890718                            2.400092</span></span>
<span><span class="co">#&gt;     Exp Ly6C_asinh -- Microglia Exp Ly6C_asinh -- Neutrophils</span></span>
<span><span class="co">#&gt;                           &lt;num&gt;                         &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:                   0.8916724                      3.963225</span></span>
<span><span class="co">#&gt;  2:                   0.8539973                      4.109470</span></span>
<span><span class="co">#&gt;  3:                   0.9004572                      4.150608</span></span>
<span><span class="co">#&gt;  4:                   0.8547482                      4.026134</span></span>
<span><span class="co">#&gt;  5:                   0.8487780                      4.111840</span></span>
<span><span class="co">#&gt;  6:                   0.9163431                      4.025573</span></span>
<span><span class="co">#&gt;  7:                   1.6985734                      3.239471</span></span>
<span><span class="co">#&gt;  8:                   1.5986071                      2.851287</span></span>
<span><span class="co">#&gt;  9:                   1.5641569                      2.484206</span></span>
<span><span class="co">#&gt; 10:                   1.6370403                      2.976742</span></span>
<span><span class="co">#&gt; 11:                   1.4260377                      2.871137</span></span>
<span><span class="co">#&gt; 12:                   1.4905523                      2.293279</span></span>
<span><span class="co">#&gt;     Exp Ly6C_asinh -- NK cells Percent of sample -- CD4 T cells</span></span>
<span><span class="co">#&gt;                          &lt;num&gt;                            &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:                  0.6460165                        0.6641844</span></span>
<span><span class="co">#&gt;  2:                  0.5383660                        0.4930263</span></span>
<span><span class="co">#&gt;  3:                  0.4378819                        0.4913660</span></span>
<span><span class="co">#&gt;  4:                  0.3142635                        0.1936165</span></span>
<span><span class="co">#&gt;  5:                  0.3423965                        0.1648654</span></span>
<span><span class="co">#&gt;  6:                  0.4254617                        0.3475772</span></span>
<span><span class="co">#&gt;  7:                  1.5854437                        5.2311978</span></span>
<span><span class="co">#&gt;  8:                  1.2815850                        1.9959342</span></span>
<span><span class="co">#&gt;  9:                  1.0971466                        2.5439090</span></span>
<span><span class="co">#&gt; 10:                  1.3549706                        2.5392563</span></span>
<span><span class="co">#&gt; 11:                  1.1519772                        3.0641718</span></span>
<span><span class="co">#&gt; 12:                  1.0961671                        3.3345763</span></span>
<span><span class="co">#&gt;     Percent of sample -- CD8 T cells Percent of sample -- Infil Macrophages</span></span>
<span><span class="co">#&gt;                                &lt;num&gt;                                  &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:                         2.797625                              2.5259133</span></span>
<span><span class="co">#&gt;  2:                         2.951670                              0.9925397</span></span>
<span><span class="co">#&gt;  3:                         2.969255                              0.8283027</span></span>
<span><span class="co">#&gt;  4:                         2.153250                              0.3402957</span></span>
<span><span class="co">#&gt;  5:                         3.388899                              0.4579593</span></span>
<span><span class="co">#&gt;  6:                         6.583521                              1.1654058</span></span>
<span><span class="co">#&gt;  7:                        62.674095                              4.6963788</span></span>
<span><span class="co">#&gt;  8:                        69.605125                              2.2731473</span></span>
<span><span class="co">#&gt;  9:                        69.841865                              3.4314645</span></span>
<span><span class="co">#&gt; 10:                        67.316561                              3.0295955</span></span>
<span><span class="co">#&gt; 11:                        65.986437                              3.5099837</span></span>
<span><span class="co">#&gt; 12:                        67.186917                              3.7287594</span></span>
<span><span class="co">#&gt;     Percent of sample -- Microglia Percent of sample -- Neutrophils</span></span>
<span><span class="co">#&gt;                              &lt;num&gt;                            &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:                       90.70142                         2.737245</span></span>
<span><span class="co">#&gt;  2:                       91.91048                         2.821927</span></span>
<span><span class="co">#&gt;  3:                       91.63976                         3.067528</span></span>
<span><span class="co">#&gt;  4:                       94.59634                         2.194321</span></span>
<span><span class="co">#&gt;  5:                       92.04983                         3.150760</span></span>
<span><span class="co">#&gt;  6:                       83.68432                         7.483132</span></span>
<span><span class="co">#&gt;  7:                       12.51253                         2.356546</span></span>
<span><span class="co">#&gt;  8:                       12.44995                         4.909752</span></span>
<span><span class="co">#&gt;  9:                       12.05075                         3.950247</span></span>
<span><span class="co">#&gt; 10:                       11.25445                         8.067247</span></span>
<span><span class="co">#&gt; 11:                       12.09343                         1.802085</span></span>
<span><span class="co">#&gt; 12:                       13.43419                         1.944282</span></span>
<span><span class="co">#&gt;     Percent of sample -- NK cells</span></span>
<span><span class="co">#&gt;                             &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:                     0.5736138</span></span>
<span><span class="co">#&gt;  2:                     0.8303600</span></span>
<span><span class="co">#&gt;  3:                     1.0037905</span></span>
<span><span class="co">#&gt;  4:                     0.5221779</span></span>
<span><span class="co">#&gt;  5:                     0.7876901</span></span>
<span><span class="co">#&gt;  6:                     0.7360458</span></span>
<span><span class="co">#&gt;  7:                    12.5292479</span></span>
<span><span class="co">#&gt;  8:                     8.7660938</span></span>
<span><span class="co">#&gt;  9:                     8.1817614</span></span>
<span><span class="co">#&gt; 10:                     7.7928901</span></span>
<span><span class="co">#&gt; 11:                    13.5438905</span></span>
<span><span class="co">#&gt; 12:                    10.3712779</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sum.dat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;       [,1]                                    </span></span>
<span><span class="co">#&gt;  [1,] "Sample"                                </span></span>
<span><span class="co">#&gt;  [2,] "Group"                                 </span></span>
<span><span class="co">#&gt;  [3,] "Batch"                                 </span></span>
<span><span class="co">#&gt;  [4,] "Cells per sample -- CD4 T cells"       </span></span>
<span><span class="co">#&gt;  [5,] "Cells per sample -- CD8 T cells"       </span></span>
<span><span class="co">#&gt;  [6,] "Cells per sample -- Infil Macrophages" </span></span>
<span><span class="co">#&gt;  [7,] "Cells per sample -- Microglia"         </span></span>
<span><span class="co">#&gt;  [8,] "Cells per sample -- Neutrophils"       </span></span>
<span><span class="co">#&gt;  [9,] "Cells per sample -- NK cells"          </span></span>
<span><span class="co">#&gt; [10,] "Exp CD11b_asinh -- CD4 T cells"        </span></span>
<span><span class="co">#&gt; [11,] "Exp CD11b_asinh -- CD8 T cells"        </span></span>
<span><span class="co">#&gt; [12,] "Exp CD11b_asinh -- Infil Macrophages"  </span></span>
<span><span class="co">#&gt; [13,] "Exp CD11b_asinh -- Microglia"          </span></span>
<span><span class="co">#&gt; [14,] "Exp CD11b_asinh -- Neutrophils"        </span></span>
<span><span class="co">#&gt; [15,] "Exp CD11b_asinh -- NK cells"           </span></span>
<span><span class="co">#&gt; [16,] "Exp Ly6C_asinh -- CD4 T cells"         </span></span>
<span><span class="co">#&gt; [17,] "Exp Ly6C_asinh -- CD8 T cells"         </span></span>
<span><span class="co">#&gt; [18,] "Exp Ly6C_asinh -- Infil Macrophages"   </span></span>
<span><span class="co">#&gt; [19,] "Exp Ly6C_asinh -- Microglia"           </span></span>
<span><span class="co">#&gt; [20,] "Exp Ly6C_asinh -- Neutrophils"         </span></span>
<span><span class="co">#&gt; [21,] "Exp Ly6C_asinh -- NK cells"            </span></span>
<span><span class="co">#&gt; [22,] "Percent of sample -- CD4 T cells"      </span></span>
<span><span class="co">#&gt; [23,] "Percent of sample -- CD8 T cells"      </span></span>
<span><span class="co">#&gt; [24,] "Percent of sample -- Infil Macrophages"</span></span>
<span><span class="co">#&gt; [25,] "Percent of sample -- Microglia"        </span></span>
<span><span class="co">#&gt; [26,] "Percent of sample -- Neutrophils"      </span></span>
<span><span class="co">#&gt; [27,] "Percent of sample -- NK cells"</span></span>
<span></span>
<span><span class="va">annot.cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">group.col</span>, <span class="va">batch.col</span><span class="op">)</span></span></code></pre></div>
<p>Specify which columns we want to plot.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">plot.cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sum.dat</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span><span class="op">:</span><span class="fl">27</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">plot.cols</span></span>
<span><span class="co">#&gt;  [1] "Cells per sample -- CD4 T cells"       </span></span>
<span><span class="co">#&gt;  [2] "Cells per sample -- CD8 T cells"       </span></span>
<span><span class="co">#&gt;  [3] "Cells per sample -- Infil Macrophages" </span></span>
<span><span class="co">#&gt;  [4] "Cells per sample -- Microglia"         </span></span>
<span><span class="co">#&gt;  [5] "Cells per sample -- Neutrophils"       </span></span>
<span><span class="co">#&gt;  [6] "Cells per sample -- NK cells"          </span></span>
<span><span class="co">#&gt;  [7] "Exp CD11b_asinh -- CD4 T cells"        </span></span>
<span><span class="co">#&gt;  [8] "Exp CD11b_asinh -- CD8 T cells"        </span></span>
<span><span class="co">#&gt;  [9] "Exp CD11b_asinh -- Infil Macrophages"  </span></span>
<span><span class="co">#&gt; [10] "Exp CD11b_asinh -- Microglia"          </span></span>
<span><span class="co">#&gt; [11] "Exp CD11b_asinh -- Neutrophils"        </span></span>
<span><span class="co">#&gt; [12] "Exp CD11b_asinh -- NK cells"           </span></span>
<span><span class="co">#&gt; [13] "Exp Ly6C_asinh -- CD4 T cells"         </span></span>
<span><span class="co">#&gt; [14] "Exp Ly6C_asinh -- CD8 T cells"         </span></span>
<span><span class="co">#&gt; [15] "Exp Ly6C_asinh -- Infil Macrophages"   </span></span>
<span><span class="co">#&gt; [16] "Exp Ly6C_asinh -- Microglia"           </span></span>
<span><span class="co">#&gt; [17] "Exp Ly6C_asinh -- Neutrophils"         </span></span>
<span><span class="co">#&gt; [18] "Exp Ly6C_asinh -- NK cells"            </span></span>
<span><span class="co">#&gt; [19] "Percent of sample -- CD4 T cells"      </span></span>
<span><span class="co">#&gt; [20] "Percent of sample -- CD8 T cells"      </span></span>
<span><span class="co">#&gt; [21] "Percent of sample -- Infil Macrophages"</span></span>
<span><span class="co">#&gt; [22] "Percent of sample -- Microglia"        </span></span>
<span><span class="co">#&gt; [23] "Percent of sample -- Neutrophils"      </span></span>
<span><span class="co">#&gt; [24] "Percent of sample -- NK cells"</span></span></code></pre></div>
<p>Reorder the data such that sample appear in the specify group
order.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### Reorder summary data and SAVE</span></span>
<span></span>
<span><span class="va">sum.dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do.reorder.html">do.reorder</a></span><span class="op">(</span><span class="va">sum.dat</span>, <span class="va">group.col</span>, <span class="va">grp.order</span><span class="op">)</span></span>
<span><span class="va">sum.dat</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt;         Sample  Group  Batch</span></span>
<span><span class="co">#&gt;         &lt;fctr&gt; &lt;fctr&gt; &lt;char&gt;</span></span>
<span><span class="co">#&gt;  1: 01_Mock_01   Mock      A</span></span>
<span><span class="co">#&gt;  2: 02_Mock_02   Mock      B</span></span>
<span><span class="co">#&gt;  3: 03_Mock_03   Mock      B</span></span>
<span><span class="co">#&gt;  4: 04_Mock_04   Mock      A</span></span>
<span><span class="co">#&gt;  5: 05_Mock_05   Mock      A</span></span>
<span><span class="co">#&gt;  6: 06_Mock_06   Mock      B</span></span>
<span><span class="co">#&gt;  7:  07_WNV_01    WNV      A</span></span>
<span><span class="co">#&gt;  8:  08_WNV_02    WNV      B</span></span>
<span><span class="co">#&gt;  9:  09_WNV_03    WNV      A</span></span>
<span><span class="co">#&gt; 10:  10_WNV_04    WNV      A</span></span>
<span><span class="co">#&gt; 11:  11_WNV_05    WNV      B</span></span>
<span><span class="co">#&gt; 12:  12_WNV_06    WNV      A</span></span>
<span></span>
<span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fwrite.html" class="external-link">fwrite</a></span><span class="op">(</span><span class="va">sum.dat</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">'sum.dat.csv'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="graphs">Graphs<a class="anchor" aria-label="anchor" href="#graphs"></a>
</h3>
<p>We can use the run.autograph function to create violin/scatter plots
with embedded statistic – one per population/measurement type.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### Autographs</span></span>
<span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="va">plot.cols</span><span class="op">)</span><span class="op">{</span></span>
<span>    </span>
<span>    <span class="va">measure</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\ --.*"</span>, <span class="st">""</span>, <span class="va">i</span><span class="op">)</span></span>
<span>    <span class="va">measure</span></span>
<span>    </span>
<span>    <span class="va">pop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"^[^--]*.-- "</span>, <span class="st">""</span>, <span class="va">i</span><span class="op">)</span></span>
<span>    <span class="va">pop</span></span>
<span>    </span>
<span>    <span class="fu"><a href="../reference/make.autograph.html">make.autograph</a></span><span class="op">(</span><span class="va">sum.dat</span>,</span>
<span>                    x.axis <span class="op">=</span> <span class="va">group.col</span>,</span>
<span>                    y.axis <span class="op">=</span> <span class="va">i</span>,</span>
<span>                    y.axis.label <span class="op">=</span> <span class="va">measure</span>,</span>
<span>                    </span>
<span>                    grp.order <span class="op">=</span> <span class="va">grp.order</span>,</span>
<span>                    my_comparisons <span class="op">=</span> <span class="va">comparisons</span>,</span>
<span>                    </span>
<span>                    Variance_test <span class="op">=</span> <span class="va">variance.test</span>,</span>
<span>                    Pairwise_test <span class="op">=</span> <span class="va">pairwise.test</span>,</span>
<span>                    </span>
<span>                    title <span class="op">=</span> <span class="va">pop</span>,</span>
<span>                    subtitle <span class="op">=</span> <span class="va">measure</span>,</span>
<span>                    filename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">i</span>, <span class="st">'.pdf'</span><span class="op">)</span>,</span>
<span>                    path <span class="op">=</span> <span class="va">outdir</span><span class="op">)</span></span>
<span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="images/simple_discovery/image2021-1-13_16-42-20.png" style="width:50.0%"></p>
<p><img src="images/simple_discovery/image2021-1-13_16-42-30.png" style="width:50.0%"></p>
</div>
<div class="section level3">
<h3 id="heatmaps">Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"></a>
</h3>
<p>We can also create a global heatmap show the z-score of each
population/measurement type against each sample.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">### Create a fold change heatmap</span></span>
<span></span>
<span><span class="co">## Z-score calculation</span></span>
<span><span class="va">sum.dat.z</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/do.zscore.html">do.zscore</a></span><span class="op">(</span><span class="va">sum.dat</span>, <span class="va">plot.cols</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Group </span></span>
<span><span class="va">t.first</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">grp.order</span>, <span class="va">sum.dat.z</span><span class="op">[[</span><span class="va">group.col</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">t.first</span> <span class="op">&lt;-</span> <span class="va">t.first</span> <span class="op">-</span><span class="fl">1</span></span>
<span><span class="va">t.first</span></span>
<span><span class="co">#&gt; [1] 0 6</span></span>
<span></span>
<span><span class="co">## Make heatmap</span></span>
<span><span class="fu"><a href="../reference/make.pheatmap.html">make.pheatmap</a></span><span class="op">(</span><span class="va">sum.dat.z</span>, </span>
<span>              sample.col <span class="op">=</span> <span class="va">sample.col</span>, </span>
<span>              plot.cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">plot.cols</span>, <span class="st">'_zscore'</span><span class="op">)</span>, </span>
<span>              is.fold <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>              plot.title <span class="op">=</span> <span class="st">'Z-score'</span>,</span>
<span>              annot.cols <span class="op">=</span> <span class="va">annot.cols</span>,</span>
<span>              dendrograms <span class="op">=</span> <span class="st">'column'</span>,</span>
<span>              row.sep <span class="op">=</span> <span class="va">t.first</span>,</span>
<span>              cutree_cols <span class="op">=</span> <span class="fl">3</span>,</span>
<span>              path <span class="op">=</span> <span class="va">outdir</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/simple_discovery/image2021-1-13_16-43-1.png" style="width:70.0%"></p>
</div>
</div>
<div class="section level2">
<h2 id="save-session-info">8. Save session info<a class="anchor" aria-label="anchor" href="#save-session-info"></a>
</h2>
<p>For the final step of our setup, we want to record the session info
our R session, and save this in a folder we’ll call “Output-info”.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### Session info and metadata</span></span>
<span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">OutputDirectory</span>, <span class="st">"Output - info"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sink.html" class="external-link">sink</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"session_info.txt"</span><span class="op">)</span>, append<span class="op">=</span><span class="cn">TRUE</span>, split<span class="op">=</span><span class="cn">FALSE</span>, type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"output"</span>, <span class="st">"message"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sink.html" class="external-link">sink</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/write.html" class="external-link">write</a></span><span class="op">(</span><span class="va">cellular.cols</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"cellular.cols.txt"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/write.html" class="external-link">write</a></span><span class="op">(</span><span class="va">cluster.cols</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"cluster.cols.txt"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thomas Ashhurst, Felix Marsh-Wakefield, Givanna Putri.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
